// kubemq-js/lib/protos/kubemq.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.kubemq = void 0;
const tslib_1 = require("tslib");
/**
 * Generated by the protoc-gen-ts.  DO NOT EDIT!
 * compiler version: 5.27.3
 * source: kubemq.proto
 * git: https://github.com/thesayyn/protoc-gen-ts */
const pb_1 = require("google-protobuf");
const grpc_1 = require("@grpc/grpc-js");
var kubemq;
(function (kubemq) {
    var _PingResult_one_of_decls, _Empty_one_of_decls, _Result_one_of_decls, _Event_one_of_decls, _EventReceive_one_of_decls, _Subscribe_one_of_decls, _Request_one_of_decls, _Response_one_of_decls, _QueueMessage_one_of_decls, _QueueMessagesBatchRequest_one_of_decls, _QueueMessagesBatchResponse_one_of_decls, _QueueMessageAttributes_one_of_decls, _QueueMessagePolicy_one_of_decls, _SendQueueMessageResult_one_of_decls, _ReceiveQueueMessagesRequest_one_of_decls, _ReceiveQueueMessagesResponse_one_of_decls, _AckAllQueueMessagesRequest_one_of_decls, _AckAllQueueMessagesResponse_one_of_decls, _StreamQueueMessagesRequest_one_of_decls, _StreamQueueMessagesResponse_one_of_decls, _QueuesUpstreamRequest_one_of_decls, _QueuesUpstreamResponse_one_of_decls, _QueuesDownstreamRequest_one_of_decls, _QueuesDownstreamResponse_one_of_decls, _QueueInfo_one_of_decls, _QueuesInfo_one_of_decls, _QueuesInfoRequest_one_of_decls, _QueuesInfoResponse_one_of_decls;
    let StreamRequestType;
    (function (StreamRequestType) {
        StreamRequestType[StreamRequestType["StreamRequestTypeUnknown"] = 0] = "StreamRequestTypeUnknown";
        StreamRequestType[StreamRequestType["ReceiveMessage"] = 1] = "ReceiveMessage";
        StreamRequestType[StreamRequestType["AckMessage"] = 2] = "AckMessage";
        StreamRequestType[StreamRequestType["RejectMessage"] = 3] = "RejectMessage";
        StreamRequestType[StreamRequestType["ModifyVisibility"] = 4] = "ModifyVisibility";
        StreamRequestType[StreamRequestType["ResendMessage"] = 5] = "ResendMessage";
        StreamRequestType[StreamRequestType["SendModifiedMessage"] = 6] = "SendModifiedMessage";
    })(StreamRequestType = kubemq.StreamRequestType || (kubemq.StreamRequestType = {}));
    let QueuesDownstreamRequestType;
    (function (QueuesDownstreamRequestType) {
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["PollRequestTypeUnknown"] = 0] = "PollRequestTypeUnknown";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["Get"] = 1] = "Get";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["AckAll"] = 2] = "AckAll";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["AckRange"] = 3] = "AckRange";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["NAckAll"] = 4] = "NAckAll";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["NAckRange"] = 5] = "NAckRange";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["ReQueueAll"] = 6] = "ReQueueAll";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["ReQueueRange"] = 7] = "ReQueueRange";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["ActiveOffsets"] = 8] = "ActiveOffsets";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["TransactionStatus"] = 9] = "TransactionStatus";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["CloseByClient"] = 10] = "CloseByClient";
        QueuesDownstreamRequestType[QueuesDownstreamRequestType["CloseByServer"] = 11] = "CloseByServer";
    })(QueuesDownstreamRequestType = kubemq.QueuesDownstreamRequestType || (kubemq.QueuesDownstreamRequestType = {}));
    class PingResult extends pb_1.Message {
        constructor(data) {
            super();
            _PingResult_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _PingResult_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("Host" in data && data.Host != undefined) {
                    this.Host = data.Host;
                }
                if ("Version" in data && data.Version != undefined) {
                    this.Version = data.Version;
                }
                if ("ServerStartTime" in data && data.ServerStartTime != undefined) {
                    this.ServerStartTime = data.ServerStartTime;
                }
                if ("ServerUpTimeSeconds" in data && data.ServerUpTimeSeconds != undefined) {
                    this.ServerUpTimeSeconds = data.ServerUpTimeSeconds;
                }
            }
        }
        get Host() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set Host(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Version() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set Version(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get ServerStartTime() {
            return pb_1.Message.getFieldWithDefault(this, 3, 0);
        }
        set ServerStartTime(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get ServerUpTimeSeconds() {
            return pb_1.Message.getFieldWithDefault(this, 4, 0);
        }
        set ServerUpTimeSeconds(value) {
            pb_1.Message.setField(this, 4, value);
        }
        static fromObject(data) {
            const message = new PingResult({});
            if (data.Host != null) {
                message.Host = data.Host;
            }
            if (data.Version != null) {
                message.Version = data.Version;
            }
            if (data.ServerStartTime != null) {
                message.ServerStartTime = data.ServerStartTime;
            }
            if (data.ServerUpTimeSeconds != null) {
                message.ServerUpTimeSeconds = data.ServerUpTimeSeconds;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.Host != null) {
                data.Host = this.Host;
            }
            if (this.Version != null) {
                data.Version = this.Version;
            }
            if (this.ServerStartTime != null) {
                data.ServerStartTime = this.ServerStartTime;
            }
            if (this.ServerUpTimeSeconds != null) {
                data.ServerUpTimeSeconds = this.ServerUpTimeSeconds;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.Host.length)
                writer.writeString(1, this.Host);
            if (this.Version.length)
                writer.writeString(2, this.Version);
            if (this.ServerStartTime != 0)
                writer.writeInt64(3, this.ServerStartTime);
            if (this.ServerUpTimeSeconds != 0)
                writer.writeInt64(4, this.ServerUpTimeSeconds);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new PingResult();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.Host = reader.readString();
                        break;
                    case 2:
                        message.Version = reader.readString();
                        break;
                    case 3:
                        message.ServerStartTime = reader.readInt64();
                        break;
                    case 4:
                        message.ServerUpTimeSeconds = reader.readInt64();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return PingResult.deserialize(bytes);
        }
    }
    _PingResult_one_of_decls = new WeakMap();
    kubemq.PingResult = PingResult;
    class Empty extends pb_1.Message {
        constructor(data) {
            super();
            _Empty_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _Empty_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") { }
        }
        static fromObject(data) {
            const message = new Empty({});
            return message;
        }
        toObject() {
            const data = {};
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new Empty();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return Empty.deserialize(bytes);
        }
    }
    _Empty_one_of_decls = new WeakMap();
    kubemq.Empty = Empty;
    class Result extends pb_1.Message {
        constructor(data) {
            super();
            _Result_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _Result_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("EventID" in data && data.EventID != undefined) {
                    this.EventID = data.EventID;
                }
                if ("Sent" in data && data.Sent != undefined) {
                    this.Sent = data.Sent;
                }
                if ("Error" in data && data.Error != undefined) {
                    this.Error = data.Error;
                }
            }
        }
        get EventID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set EventID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Sent() {
            return pb_1.Message.getFieldWithDefault(this, 2, false);
        }
        set Sent(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Error() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set Error(value) {
            pb_1.Message.setField(this, 3, value);
        }
        static fromObject(data) {
            const message = new Result({});
            if (data.EventID != null) {
                message.EventID = data.EventID;
            }
            if (data.Sent != null) {
                message.Sent = data.Sent;
            }
            if (data.Error != null) {
                message.Error = data.Error;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.EventID != null) {
                data.EventID = this.EventID;
            }
            if (this.Sent != null) {
                data.Sent = this.Sent;
            }
            if (this.Error != null) {
                data.Error = this.Error;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.EventID.length)
                writer.writeString(1, this.EventID);
            if (this.Sent != false)
                writer.writeBool(2, this.Sent);
            if (this.Error.length)
                writer.writeString(3, this.Error);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new Result();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.EventID = reader.readString();
                        break;
                    case 2:
                        message.Sent = reader.readBool();
                        break;
                    case 3:
                        message.Error = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return Result.deserialize(bytes);
        }
    }
    _Result_one_of_decls = new WeakMap();
    kubemq.Result = Result;
    class Event extends pb_1.Message {
        constructor(data) {
            super();
            _Event_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _Event_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("EventID" in data && data.EventID != undefined) {
                    this.EventID = data.EventID;
                }
                if ("ClientID" in data && data.ClientID != undefined) {
                    this.ClientID = data.ClientID;
                }
                if ("Channel" in data && data.Channel != undefined) {
                    this.Channel = data.Channel;
                }
                if ("Metadata" in data && data.Metadata != undefined) {
                    this.Metadata = data.Metadata;
                }
                if ("Body" in data && data.Body != undefined) {
                    this.Body = data.Body;
                }
                if ("Store" in data && data.Store != undefined) {
                    this.Store = data.Store;
                }
                if ("Tags" in data && data.Tags != undefined) {
                    this.Tags = data.Tags;
                }
            }
            if (!this.Tags)
                this.Tags = new Map();
        }
        get EventID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set EventID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get ClientID() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set ClientID(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Channel() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set Channel(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Metadata() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set Metadata(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get Body() {
            return pb_1.Message.getFieldWithDefault(this, 5, new Uint8Array(0));
        }
        set Body(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get Store() {
            return pb_1.Message.getFieldWithDefault(this, 6, false);
        }
        set Store(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get Tags() {
            return pb_1.Message.getField(this, 7);
        }
        set Tags(value) {
            pb_1.Message.setField(this, 7, value);
        }
        static fromObject(data) {
            const message = new Event({});
            if (data.EventID != null) {
                message.EventID = data.EventID;
            }
            if (data.ClientID != null) {
                message.ClientID = data.ClientID;
            }
            if (data.Channel != null) {
                message.Channel = data.Channel;
            }
            if (data.Metadata != null) {
                message.Metadata = data.Metadata;
            }
            if (data.Body != null) {
                message.Body = data.Body;
            }
            if (data.Store != null) {
                message.Store = data.Store;
            }
            if (typeof data.Tags == "object") {
                message.Tags = new Map(Object.entries(data.Tags));
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.EventID != null) {
                data.EventID = this.EventID;
            }
            if (this.ClientID != null) {
                data.ClientID = this.ClientID;
            }
            if (this.Channel != null) {
                data.Channel = this.Channel;
            }
            if (this.Metadata != null) {
                data.Metadata = this.Metadata;
            }
            if (this.Body != null) {
                data.Body = this.Body;
            }
            if (this.Store != null) {
                data.Store = this.Store;
            }
            if (this.Tags != null) {
                data.Tags = (Object.fromEntries)(this.Tags);
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.EventID.length)
                writer.writeString(1, this.EventID);
            if (this.ClientID.length)
                writer.writeString(2, this.ClientID);
            if (this.Channel.length)
                writer.writeString(3, this.Channel);
            if (this.Metadata.length)
                writer.writeString(4, this.Metadata);
            if (this.Body.length)
                writer.writeBytes(5, this.Body);
            if (this.Store != false)
                writer.writeBool(6, this.Store);
            for (const [key, value] of this.Tags) {
                writer.writeMessage(7, this.Tags, () => {
                    writer.writeString(1, key);
                    writer.writeString(2, value);
                });
            }
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new Event();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.EventID = reader.readString();
                        break;
                    case 2:
                        message.ClientID = reader.readString();
                        break;
                    case 3:
                        message.Channel = reader.readString();
                        break;
                    case 4:
                        message.Metadata = reader.readString();
                        break;
                    case 5:
                        message.Body = reader.readBytes();
                        break;
                    case 6:
                        message.Store = reader.readBool();
                        break;
                    case 7:
                        reader.readMessage(message, () => pb_1.Map.deserializeBinary(message.Tags, reader, reader.readString, reader.readString));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return Event.deserialize(bytes);
        }
    }
    _Event_one_of_decls = new WeakMap();
    kubemq.Event = Event;
    class EventReceive extends pb_1.Message {
        constructor(data) {
            super();
            _EventReceive_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _EventReceive_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("EventID" in data && data.EventID != undefined) {
                    this.EventID = data.EventID;
                }
                if ("Channel" in data && data.Channel != undefined) {
                    this.Channel = data.Channel;
                }
                if ("Metadata" in data && data.Metadata != undefined) {
                    this.Metadata = data.Metadata;
                }
                if ("Body" in data && data.Body != undefined) {
                    this.Body = data.Body;
                }
                if ("Timestamp" in data && data.Timestamp != undefined) {
                    this.Timestamp = data.Timestamp;
                }
                if ("Sequence" in data && data.Sequence != undefined) {
                    this.Sequence = data.Sequence;
                }
                if ("Tags" in data && data.Tags != undefined) {
                    this.Tags = data.Tags;
                }
            }
            if (!this.Tags)
                this.Tags = new Map();
        }
        get EventID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set EventID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Channel() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set Channel(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Metadata() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set Metadata(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Body() {
            return pb_1.Message.getFieldWithDefault(this, 4, new Uint8Array(0));
        }
        set Body(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get Timestamp() {
            return pb_1.Message.getFieldWithDefault(this, 5, 0);
        }
        set Timestamp(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get Sequence() {
            return pb_1.Message.getFieldWithDefault(this, 6, 0);
        }
        set Sequence(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get Tags() {
            return pb_1.Message.getField(this, 7);
        }
        set Tags(value) {
            pb_1.Message.setField(this, 7, value);
        }
        static fromObject(data) {
            const message = new EventReceive({});
            if (data.EventID != null) {
                message.EventID = data.EventID;
            }
            if (data.Channel != null) {
                message.Channel = data.Channel;
            }
            if (data.Metadata != null) {
                message.Metadata = data.Metadata;
            }
            if (data.Body != null) {
                message.Body = data.Body;
            }
            if (data.Timestamp != null) {
                message.Timestamp = data.Timestamp;
            }
            if (data.Sequence != null) {
                message.Sequence = data.Sequence;
            }
            if (typeof data.Tags == "object") {
                message.Tags = new Map(Object.entries(data.Tags));
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.EventID != null) {
                data.EventID = this.EventID;
            }
            if (this.Channel != null) {
                data.Channel = this.Channel;
            }
            if (this.Metadata != null) {
                data.Metadata = this.Metadata;
            }
            if (this.Body != null) {
                data.Body = this.Body;
            }
            if (this.Timestamp != null) {
                data.Timestamp = this.Timestamp;
            }
            if (this.Sequence != null) {
                data.Sequence = this.Sequence;
            }
            if (this.Tags != null) {
                data.Tags = (Object.fromEntries)(this.Tags);
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.EventID.length)
                writer.writeString(1, this.EventID);
            if (this.Channel.length)
                writer.writeString(2, this.Channel);
            if (this.Metadata.length)
                writer.writeString(3, this.Metadata);
            if (this.Body.length)
                writer.writeBytes(4, this.Body);
            if (this.Timestamp != 0)
                writer.writeInt64(5, this.Timestamp);
            if (this.Sequence != 0)
                writer.writeUint64(6, this.Sequence);
            for (const [key, value] of this.Tags) {
                writer.writeMessage(7, this.Tags, () => {
                    writer.writeString(1, key);
                    writer.writeString(2, value);
                });
            }
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new EventReceive();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.EventID = reader.readString();
                        break;
                    case 2:
                        message.Channel = reader.readString();
                        break;
                    case 3:
                        message.Metadata = reader.readString();
                        break;
                    case 4:
                        message.Body = reader.readBytes();
                        break;
                    case 5:
                        message.Timestamp = reader.readInt64();
                        break;
                    case 6:
                        message.Sequence = reader.readUint64();
                        break;
                    case 7:
                        reader.readMessage(message, () => pb_1.Map.deserializeBinary(message.Tags, reader, reader.readString, reader.readString));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return EventReceive.deserialize(bytes);
        }
    }
    _EventReceive_one_of_decls = new WeakMap();
    kubemq.EventReceive = EventReceive;
    class Subscribe extends pb_1.Message {
        constructor(data) {
            super();
            _Subscribe_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _Subscribe_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("SubscribeTypeData" in data && data.SubscribeTypeData != undefined) {
                    this.SubscribeTypeData = data.SubscribeTypeData;
                }
                if ("ClientID" in data && data.ClientID != undefined) {
                    this.ClientID = data.ClientID;
                }
                if ("Channel" in data && data.Channel != undefined) {
                    this.Channel = data.Channel;
                }
                if ("Group" in data && data.Group != undefined) {
                    this.Group = data.Group;
                }
                if ("EventsStoreTypeData" in data && data.EventsStoreTypeData != undefined) {
                    this.EventsStoreTypeData = data.EventsStoreTypeData;
                }
                if ("EventsStoreTypeValue" in data && data.EventsStoreTypeValue != undefined) {
                    this.EventsStoreTypeValue = data.EventsStoreTypeValue;
                }
            }
        }
        get SubscribeTypeData() {
            return pb_1.Message.getFieldWithDefault(this, 1, Subscribe.SubscribeType.SubscribeTypeUndefined);
        }
        set SubscribeTypeData(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get ClientID() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set ClientID(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Channel() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set Channel(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Group() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set Group(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get EventsStoreTypeData() {
            return pb_1.Message.getFieldWithDefault(this, 5, Subscribe.EventsStoreType.EventsStoreTypeUndefined);
        }
        set EventsStoreTypeData(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get EventsStoreTypeValue() {
            return pb_1.Message.getFieldWithDefault(this, 6, 0);
        }
        set EventsStoreTypeValue(value) {
            pb_1.Message.setField(this, 6, value);
        }
        static fromObject(data) {
            const message = new Subscribe({});
            if (data.SubscribeTypeData != null) {
                message.SubscribeTypeData = data.SubscribeTypeData;
            }
            if (data.ClientID != null) {
                message.ClientID = data.ClientID;
            }
            if (data.Channel != null) {
                message.Channel = data.Channel;
            }
            if (data.Group != null) {
                message.Group = data.Group;
            }
            if (data.EventsStoreTypeData != null) {
                message.EventsStoreTypeData = data.EventsStoreTypeData;
            }
            if (data.EventsStoreTypeValue != null) {
                message.EventsStoreTypeValue = data.EventsStoreTypeValue;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.SubscribeTypeData != null) {
                data.SubscribeTypeData = this.SubscribeTypeData;
            }
            if (this.ClientID != null) {
                data.ClientID = this.ClientID;
            }
            if (this.Channel != null) {
                data.Channel = this.Channel;
            }
            if (this.Group != null) {
                data.Group = this.Group;
            }
            if (this.EventsStoreTypeData != null) {
                data.EventsStoreTypeData = this.EventsStoreTypeData;
            }
            if (this.EventsStoreTypeValue != null) {
                data.EventsStoreTypeValue = this.EventsStoreTypeValue;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.SubscribeTypeData != Subscribe.SubscribeType.SubscribeTypeUndefined)
                writer.writeEnum(1, this.SubscribeTypeData);
            if (this.ClientID.length)
                writer.writeString(2, this.ClientID);
            if (this.Channel.length)
                writer.writeString(3, this.Channel);
            if (this.Group.length)
                writer.writeString(4, this.Group);
            if (this.EventsStoreTypeData != Subscribe.EventsStoreType.EventsStoreTypeUndefined)
                writer.writeEnum(5, this.EventsStoreTypeData);
            if (this.EventsStoreTypeValue != 0)
                writer.writeInt64(6, this.EventsStoreTypeValue);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new Subscribe();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.SubscribeTypeData = reader.readEnum();
                        break;
                    case 2:
                        message.ClientID = reader.readString();
                        break;
                    case 3:
                        message.Channel = reader.readString();
                        break;
                    case 4:
                        message.Group = reader.readString();
                        break;
                    case 5:
                        message.EventsStoreTypeData = reader.readEnum();
                        break;
                    case 6:
                        message.EventsStoreTypeValue = reader.readInt64();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return Subscribe.deserialize(bytes);
        }
    }
    _Subscribe_one_of_decls = new WeakMap();
    kubemq.Subscribe = Subscribe;
    (function (Subscribe) {
        let SubscribeType;
        (function (SubscribeType) {
            SubscribeType[SubscribeType["SubscribeTypeUndefined"] = 0] = "SubscribeTypeUndefined";
            SubscribeType[SubscribeType["Events"] = 1] = "Events";
            SubscribeType[SubscribeType["EventsStore"] = 2] = "EventsStore";
            SubscribeType[SubscribeType["Commands"] = 3] = "Commands";
            SubscribeType[SubscribeType["Queries"] = 4] = "Queries";
        })(SubscribeType = Subscribe.SubscribeType || (Subscribe.SubscribeType = {}));
        let EventsStoreType;
        (function (EventsStoreType) {
            EventsStoreType[EventsStoreType["EventsStoreTypeUndefined"] = 0] = "EventsStoreTypeUndefined";
            EventsStoreType[EventsStoreType["StartNewOnly"] = 1] = "StartNewOnly";
            EventsStoreType[EventsStoreType["StartFromFirst"] = 2] = "StartFromFirst";
            EventsStoreType[EventsStoreType["StartFromLast"] = 3] = "StartFromLast";
            EventsStoreType[EventsStoreType["StartAtSequence"] = 4] = "StartAtSequence";
            EventsStoreType[EventsStoreType["StartAtTime"] = 5] = "StartAtTime";
            EventsStoreType[EventsStoreType["StartAtTimeDelta"] = 6] = "StartAtTimeDelta";
        })(EventsStoreType = Subscribe.EventsStoreType || (Subscribe.EventsStoreType = {}));
    })(Subscribe = kubemq.Subscribe || (kubemq.Subscribe = {}));
    class Request extends pb_1.Message {
        constructor(data) {
            super();
            _Request_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _Request_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("RequestTypeData" in data && data.RequestTypeData != undefined) {
                    this.RequestTypeData = data.RequestTypeData;
                }
                if ("ClientID" in data && data.ClientID != undefined) {
                    this.ClientID = data.ClientID;
                }
                if ("Channel" in data && data.Channel != undefined) {
                    this.Channel = data.Channel;
                }
                if ("Metadata" in data && data.Metadata != undefined) {
                    this.Metadata = data.Metadata;
                }
                if ("Body" in data && data.Body != undefined) {
                    this.Body = data.Body;
                }
                if ("ReplyChannel" in data && data.ReplyChannel != undefined) {
                    this.ReplyChannel = data.ReplyChannel;
                }
                if ("Timeout" in data && data.Timeout != undefined) {
                    this.Timeout = data.Timeout;
                }
                if ("CacheKey" in data && data.CacheKey != undefined) {
                    this.CacheKey = data.CacheKey;
                }
                if ("CacheTTL" in data && data.CacheTTL != undefined) {
                    this.CacheTTL = data.CacheTTL;
                }
                if ("Span" in data && data.Span != undefined) {
                    this.Span = data.Span;
                }
                if ("Tags" in data && data.Tags != undefined) {
                    this.Tags = data.Tags;
                }
            }
            if (!this.Tags)
                this.Tags = new Map();
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get RequestTypeData() {
            return pb_1.Message.getFieldWithDefault(this, 2, Request.RequestType.RequestTypeUnknown);
        }
        set RequestTypeData(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get ClientID() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set ClientID(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Channel() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set Channel(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get Metadata() {
            return pb_1.Message.getFieldWithDefault(this, 5, "");
        }
        set Metadata(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get Body() {
            return pb_1.Message.getFieldWithDefault(this, 6, new Uint8Array(0));
        }
        set Body(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get ReplyChannel() {
            return pb_1.Message.getFieldWithDefault(this, 7, "");
        }
        set ReplyChannel(value) {
            pb_1.Message.setField(this, 7, value);
        }
        get Timeout() {
            return pb_1.Message.getFieldWithDefault(this, 8, 0);
        }
        set Timeout(value) {
            pb_1.Message.setField(this, 8, value);
        }
        get CacheKey() {
            return pb_1.Message.getFieldWithDefault(this, 9, "");
        }
        set CacheKey(value) {
            pb_1.Message.setField(this, 9, value);
        }
        get CacheTTL() {
            return pb_1.Message.getFieldWithDefault(this, 10, 0);
        }
        set CacheTTL(value) {
            pb_1.Message.setField(this, 10, value);
        }
        get Span() {
            return pb_1.Message.getFieldWithDefault(this, 11, new Uint8Array(0));
        }
        set Span(value) {
            pb_1.Message.setField(this, 11, value);
        }
        get Tags() {
            return pb_1.Message.getField(this, 12);
        }
        set Tags(value) {
            pb_1.Message.setField(this, 12, value);
        }
        static fromObject(data) {
            const message = new Request({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.RequestTypeData != null) {
                message.RequestTypeData = data.RequestTypeData;
            }
            if (data.ClientID != null) {
                message.ClientID = data.ClientID;
            }
            if (data.Channel != null) {
                message.Channel = data.Channel;
            }
            if (data.Metadata != null) {
                message.Metadata = data.Metadata;
            }
            if (data.Body != null) {
                message.Body = data.Body;
            }
            if (data.ReplyChannel != null) {
                message.ReplyChannel = data.ReplyChannel;
            }
            if (data.Timeout != null) {
                message.Timeout = data.Timeout;
            }
            if (data.CacheKey != null) {
                message.CacheKey = data.CacheKey;
            }
            if (data.CacheTTL != null) {
                message.CacheTTL = data.CacheTTL;
            }
            if (data.Span != null) {
                message.Span = data.Span;
            }
            if (typeof data.Tags == "object") {
                message.Tags = new Map(Object.entries(data.Tags));
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.RequestTypeData != null) {
                data.RequestTypeData = this.RequestTypeData;
            }
            if (this.ClientID != null) {
                data.ClientID = this.ClientID;
            }
            if (this.Channel != null) {
                data.Channel = this.Channel;
            }
            if (this.Metadata != null) {
                data.Metadata = this.Metadata;
            }
            if (this.Body != null) {
                data.Body = this.Body;
            }
            if (this.ReplyChannel != null) {
                data.ReplyChannel = this.ReplyChannel;
            }
            if (this.Timeout != null) {
                data.Timeout = this.Timeout;
            }
            if (this.CacheKey != null) {
                data.CacheKey = this.CacheKey;
            }
            if (this.CacheTTL != null) {
                data.CacheTTL = this.CacheTTL;
            }
            if (this.Span != null) {
                data.Span = this.Span;
            }
            if (this.Tags != null) {
                data.Tags = (Object.fromEntries)(this.Tags);
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.RequestTypeData != Request.RequestType.RequestTypeUnknown)
                writer.writeEnum(2, this.RequestTypeData);
            if (this.ClientID.length)
                writer.writeString(3, this.ClientID);
            if (this.Channel.length)
                writer.writeString(4, this.Channel);
            if (this.Metadata.length)
                writer.writeString(5, this.Metadata);
            if (this.Body.length)
                writer.writeBytes(6, this.Body);
            if (this.ReplyChannel.length)
                writer.writeString(7, this.ReplyChannel);
            if (this.Timeout != 0)
                writer.writeInt32(8, this.Timeout);
            if (this.CacheKey.length)
                writer.writeString(9, this.CacheKey);
            if (this.CacheTTL != 0)
                writer.writeInt32(10, this.CacheTTL);
            if (this.Span.length)
                writer.writeBytes(11, this.Span);
            for (const [key, value] of this.Tags) {
                writer.writeMessage(12, this.Tags, () => {
                    writer.writeString(1, key);
                    writer.writeString(2, value);
                });
            }
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new Request();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        message.RequestTypeData = reader.readEnum();
                        break;
                    case 3:
                        message.ClientID = reader.readString();
                        break;
                    case 4:
                        message.Channel = reader.readString();
                        break;
                    case 5:
                        message.Metadata = reader.readString();
                        break;
                    case 6:
                        message.Body = reader.readBytes();
                        break;
                    case 7:
                        message.ReplyChannel = reader.readString();
                        break;
                    case 8:
                        message.Timeout = reader.readInt32();
                        break;
                    case 9:
                        message.CacheKey = reader.readString();
                        break;
                    case 10:
                        message.CacheTTL = reader.readInt32();
                        break;
                    case 11:
                        message.Span = reader.readBytes();
                        break;
                    case 12:
                        reader.readMessage(message, () => pb_1.Map.deserializeBinary(message.Tags, reader, reader.readString, reader.readString));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return Request.deserialize(bytes);
        }
    }
    _Request_one_of_decls = new WeakMap();
    kubemq.Request = Request;
    (function (Request) {
        let RequestType;
        (function (RequestType) {
            RequestType[RequestType["RequestTypeUnknown"] = 0] = "RequestTypeUnknown";
            RequestType[RequestType["Command"] = 1] = "Command";
            RequestType[RequestType["Query"] = 2] = "Query";
        })(RequestType = Request.RequestType || (Request.RequestType = {}));
    })(Request = kubemq.Request || (kubemq.Request = {}));
    class Response extends pb_1.Message {
        constructor(data) {
            super();
            _Response_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _Response_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("ClientID" in data && data.ClientID != undefined) {
                    this.ClientID = data.ClientID;
                }
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("ReplyChannel" in data && data.ReplyChannel != undefined) {
                    this.ReplyChannel = data.ReplyChannel;
                }
                if ("Metadata" in data && data.Metadata != undefined) {
                    this.Metadata = data.Metadata;
                }
                if ("Body" in data && data.Body != undefined) {
                    this.Body = data.Body;
                }
                if ("CacheHit" in data && data.CacheHit != undefined) {
                    this.CacheHit = data.CacheHit;
                }
                if ("Timestamp" in data && data.Timestamp != undefined) {
                    this.Timestamp = data.Timestamp;
                }
                if ("Executed" in data && data.Executed != undefined) {
                    this.Executed = data.Executed;
                }
                if ("Error" in data && data.Error != undefined) {
                    this.Error = data.Error;
                }
                if ("Span" in data && data.Span != undefined) {
                    this.Span = data.Span;
                }
                if ("Tags" in data && data.Tags != undefined) {
                    this.Tags = data.Tags;
                }
            }
            if (!this.Tags)
                this.Tags = new Map();
        }
        get ClientID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set ClientID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get ReplyChannel() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set ReplyChannel(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Metadata() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set Metadata(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get Body() {
            return pb_1.Message.getFieldWithDefault(this, 5, new Uint8Array(0));
        }
        set Body(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get CacheHit() {
            return pb_1.Message.getFieldWithDefault(this, 6, false);
        }
        set CacheHit(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get Timestamp() {
            return pb_1.Message.getFieldWithDefault(this, 7, 0);
        }
        set Timestamp(value) {
            pb_1.Message.setField(this, 7, value);
        }
        get Executed() {
            return pb_1.Message.getFieldWithDefault(this, 8, false);
        }
        set Executed(value) {
            pb_1.Message.setField(this, 8, value);
        }
        get Error() {
            return pb_1.Message.getFieldWithDefault(this, 9, "");
        }
        set Error(value) {
            pb_1.Message.setField(this, 9, value);
        }
        get Span() {
            return pb_1.Message.getFieldWithDefault(this, 10, new Uint8Array(0));
        }
        set Span(value) {
            pb_1.Message.setField(this, 10, value);
        }
        get Tags() {
            return pb_1.Message.getField(this, 11);
        }
        set Tags(value) {
            pb_1.Message.setField(this, 11, value);
        }
        static fromObject(data) {
            const message = new Response({});
            if (data.ClientID != null) {
                message.ClientID = data.ClientID;
            }
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.ReplyChannel != null) {
                message.ReplyChannel = data.ReplyChannel;
            }
            if (data.Metadata != null) {
                message.Metadata = data.Metadata;
            }
            if (data.Body != null) {
                message.Body = data.Body;
            }
            if (data.CacheHit != null) {
                message.CacheHit = data.CacheHit;
            }
            if (data.Timestamp != null) {
                message.Timestamp = data.Timestamp;
            }
            if (data.Executed != null) {
                message.Executed = data.Executed;
            }
            if (data.Error != null) {
                message.Error = data.Error;
            }
            if (data.Span != null) {
                message.Span = data.Span;
            }
            if (typeof data.Tags == "object") {
                message.Tags = new Map(Object.entries(data.Tags));
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.ClientID != null) {
                data.ClientID = this.ClientID;
            }
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.ReplyChannel != null) {
                data.ReplyChannel = this.ReplyChannel;
            }
            if (this.Metadata != null) {
                data.Metadata = this.Metadata;
            }
            if (this.Body != null) {
                data.Body = this.Body;
            }
            if (this.CacheHit != null) {
                data.CacheHit = this.CacheHit;
            }
            if (this.Timestamp != null) {
                data.Timestamp = this.Timestamp;
            }
            if (this.Executed != null) {
                data.Executed = this.Executed;
            }
            if (this.Error != null) {
                data.Error = this.Error;
            }
            if (this.Span != null) {
                data.Span = this.Span;
            }
            if (this.Tags != null) {
                data.Tags = (Object.fromEntries)(this.Tags);
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.ClientID.length)
                writer.writeString(1, this.ClientID);
            if (this.RequestID.length)
                writer.writeString(2, this.RequestID);
            if (this.ReplyChannel.length)
                writer.writeString(3, this.ReplyChannel);
            if (this.Metadata.length)
                writer.writeString(4, this.Metadata);
            if (this.Body.length)
                writer.writeBytes(5, this.Body);
            if (this.CacheHit != false)
                writer.writeBool(6, this.CacheHit);
            if (this.Timestamp != 0)
                writer.writeInt64(7, this.Timestamp);
            if (this.Executed != false)
                writer.writeBool(8, this.Executed);
            if (this.Error.length)
                writer.writeString(9, this.Error);
            if (this.Span.length)
                writer.writeBytes(10, this.Span);
            for (const [key, value] of this.Tags) {
                writer.writeMessage(11, this.Tags, () => {
                    writer.writeString(1, key);
                    writer.writeString(2, value);
                });
            }
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new Response();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.ClientID = reader.readString();
                        break;
                    case 2:
                        message.RequestID = reader.readString();
                        break;
                    case 3:
                        message.ReplyChannel = reader.readString();
                        break;
                    case 4:
                        message.Metadata = reader.readString();
                        break;
                    case 5:
                        message.Body = reader.readBytes();
                        break;
                    case 6:
                        message.CacheHit = reader.readBool();
                        break;
                    case 7:
                        message.Timestamp = reader.readInt64();
                        break;
                    case 8:
                        message.Executed = reader.readBool();
                        break;
                    case 9:
                        message.Error = reader.readString();
                        break;
                    case 10:
                        message.Span = reader.readBytes();
                        break;
                    case 11:
                        reader.readMessage(message, () => pb_1.Map.deserializeBinary(message.Tags, reader, reader.readString, reader.readString));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return Response.deserialize(bytes);
        }
    }
    _Response_one_of_decls = new WeakMap();
    kubemq.Response = Response;
    class QueueMessage extends pb_1.Message {
        constructor(data) {
            super();
            _QueueMessage_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _QueueMessage_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("MessageID" in data && data.MessageID != undefined) {
                    this.MessageID = data.MessageID;
                }
                if ("ClientID" in data && data.ClientID != undefined) {
                    this.ClientID = data.ClientID;
                }
                if ("Channel" in data && data.Channel != undefined) {
                    this.Channel = data.Channel;
                }
                if ("Metadata" in data && data.Metadata != undefined) {
                    this.Metadata = data.Metadata;
                }
                if ("Body" in data && data.Body != undefined) {
                    this.Body = data.Body;
                }
                if ("Tags" in data && data.Tags != undefined) {
                    this.Tags = data.Tags;
                }
                if ("Attributes" in data && data.Attributes != undefined) {
                    this.Attributes = data.Attributes;
                }
                if ("Policy" in data && data.Policy != undefined) {
                    this.Policy = data.Policy;
                }
                if ("Topic" in data && data.Topic != undefined) {
                    this.Topic = data.Topic;
                }
                if ("Partition" in data && data.Partition != undefined) {
                    this.Partition = data.Partition;
                }
                if ("PartitionKey" in data && data.PartitionKey != undefined) {
                    this.PartitionKey = data.PartitionKey;
                }
            }
            if (!this.Tags)
                this.Tags = new Map();
        }
        get MessageID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set MessageID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get ClientID() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set ClientID(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Channel() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set Channel(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Metadata() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set Metadata(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get Body() {
            return pb_1.Message.getFieldWithDefault(this, 5, new Uint8Array(0));
        }
        set Body(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get Tags() {
            return pb_1.Message.getField(this, 6);
        }
        set Tags(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get Attributes() {
            return pb_1.Message.getWrapperField(this, QueueMessageAttributes, 7);
        }
        set Attributes(value) {
            pb_1.Message.setWrapperField(this, 7, value);
        }
        get has_Attributes() {
            return pb_1.Message.getField(this, 7) != null;
        }
        get Policy() {
            return pb_1.Message.getWrapperField(this, QueueMessagePolicy, 8);
        }
        set Policy(value) {
            pb_1.Message.setWrapperField(this, 8, value);
        }
        get has_Policy() {
            return pb_1.Message.getField(this, 8) != null;
        }
        get Topic() {
            return pb_1.Message.getFieldWithDefault(this, 9, "");
        }
        set Topic(value) {
            pb_1.Message.setField(this, 9, value);
        }
        get Partition() {
            return pb_1.Message.getFieldWithDefault(this, 10, 0);
        }
        set Partition(value) {
            pb_1.Message.setField(this, 10, value);
        }
        get PartitionKey() {
            return pb_1.Message.getFieldWithDefault(this, 11, "");
        }
        set PartitionKey(value) {
            pb_1.Message.setField(this, 11, value);
        }
        static fromObject(data) {
            const message = new QueueMessage({});
            if (data.MessageID != null) {
                message.MessageID = data.MessageID;
            }
            if (data.ClientID != null) {
                message.ClientID = data.ClientID;
            }
            if (data.Channel != null) {
                message.Channel = data.Channel;
            }
            if (data.Metadata != null) {
                message.Metadata = data.Metadata;
            }
            if (data.Body != null) {
                message.Body = data.Body;
            }
            if (typeof data.Tags == "object") {
                message.Tags = new Map(Object.entries(data.Tags));
            }
            if (data.Attributes != null) {
                message.Attributes = QueueMessageAttributes.fromObject(data.Attributes);
            }
            if (data.Policy != null) {
                message.Policy = QueueMessagePolicy.fromObject(data.Policy);
            }
            if (data.Topic != null) {
                message.Topic = data.Topic;
            }
            if (data.Partition != null) {
                message.Partition = data.Partition;
            }
            if (data.PartitionKey != null) {
                message.PartitionKey = data.PartitionKey;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.MessageID != null) {
                data.MessageID = this.MessageID;
            }
            if (this.ClientID != null) {
                data.ClientID = this.ClientID;
            }
            if (this.Channel != null) {
                data.Channel = this.Channel;
            }
            if (this.Metadata != null) {
                data.Metadata = this.Metadata;
            }
            if (this.Body != null) {
                data.Body = this.Body;
            }
            if (this.Tags != null) {
                data.Tags = (Object.fromEntries)(this.Tags);
            }
            if (this.Attributes != null) {
                data.Attributes = this.Attributes.toObject();
            }
            if (this.Policy != null) {
                data.Policy = this.Policy.toObject();
            }
            if (this.Topic != null) {
                data.Topic = this.Topic;
            }
            if (this.Partition != null) {
                data.Partition = this.Partition;
            }
            if (this.PartitionKey != null) {
                data.PartitionKey = this.PartitionKey;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.MessageID.length)
                writer.writeString(1, this.MessageID);
            if (this.ClientID.length)
                writer.writeString(2, this.ClientID);
            if (this.Channel.length)
                writer.writeString(3, this.Channel);
            if (this.Metadata.length)
                writer.writeString(4, this.Metadata);
            if (this.Body.length)
                writer.writeBytes(5, this.Body);
            for (const [key, value] of this.Tags) {
                writer.writeMessage(6, this.Tags, () => {
                    writer.writeString(1, key);
                    writer.writeString(2, value);
                });
            }
            if (this.has_Attributes)
                writer.writeMessage(7, this.Attributes, () => this.Attributes.serialize(writer));
            if (this.has_Policy)
                writer.writeMessage(8, this.Policy, () => this.Policy.serialize(writer));
            if (this.Topic.length)
                writer.writeString(9, this.Topic);
            if (this.Partition != 0)
                writer.writeInt32(10, this.Partition);
            if (this.PartitionKey.length)
                writer.writeString(11, this.PartitionKey);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueueMessage();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.MessageID = reader.readString();
                        break;
                    case 2:
                        message.ClientID = reader.readString();
                        break;
                    case 3:
                        message.Channel = reader.readString();
                        break;
                    case 4:
                        message.Metadata = reader.readString();
                        break;
                    case 5:
                        message.Body = reader.readBytes();
                        break;
                    case 6:
                        reader.readMessage(message, () => pb_1.Map.deserializeBinary(message.Tags, reader, reader.readString, reader.readString));
                        break;
                    case 7:
                        reader.readMessage(message.Attributes, () => message.Attributes = QueueMessageAttributes.deserialize(reader));
                        break;
                    case 8:
                        reader.readMessage(message.Policy, () => message.Policy = QueueMessagePolicy.deserialize(reader));
                        break;
                    case 9:
                        message.Topic = reader.readString();
                        break;
                    case 10:
                        message.Partition = reader.readInt32();
                        break;
                    case 11:
                        message.PartitionKey = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueueMessage.deserialize(bytes);
        }
    }
    _QueueMessage_one_of_decls = new WeakMap();
    kubemq.QueueMessage = QueueMessage;
    class QueueMessagesBatchRequest extends pb_1.Message {
        constructor(data) {
            super();
            _QueueMessagesBatchRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [2], tslib_1.__classPrivateFieldGet(this, _QueueMessagesBatchRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("BatchID" in data && data.BatchID != undefined) {
                    this.BatchID = data.BatchID;
                }
                if ("Messages" in data && data.Messages != undefined) {
                    this.Messages = data.Messages;
                }
            }
        }
        get BatchID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set BatchID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Messages() {
            return pb_1.Message.getRepeatedWrapperField(this, QueueMessage, 2);
        }
        set Messages(value) {
            pb_1.Message.setRepeatedWrapperField(this, 2, value);
        }
        static fromObject(data) {
            const message = new QueueMessagesBatchRequest({});
            if (data.BatchID != null) {
                message.BatchID = data.BatchID;
            }
            if (data.Messages != null) {
                message.Messages = data.Messages.map(item => QueueMessage.fromObject(item));
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.BatchID != null) {
                data.BatchID = this.BatchID;
            }
            if (this.Messages != null) {
                data.Messages = this.Messages.map((item) => item.toObject());
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.BatchID.length)
                writer.writeString(1, this.BatchID);
            if (this.Messages.length)
                writer.writeRepeatedMessage(2, this.Messages, (item) => item.serialize(writer));
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueueMessagesBatchRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.BatchID = reader.readString();
                        break;
                    case 2:
                        reader.readMessage(message.Messages, () => pb_1.Message.addToRepeatedWrapperField(message, 2, QueueMessage.deserialize(reader), QueueMessage));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueueMessagesBatchRequest.deserialize(bytes);
        }
    }
    _QueueMessagesBatchRequest_one_of_decls = new WeakMap();
    kubemq.QueueMessagesBatchRequest = QueueMessagesBatchRequest;
    class QueueMessagesBatchResponse extends pb_1.Message {
        constructor(data) {
            super();
            _QueueMessagesBatchResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [2], tslib_1.__classPrivateFieldGet(this, _QueueMessagesBatchResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("BatchID" in data && data.BatchID != undefined) {
                    this.BatchID = data.BatchID;
                }
                if ("Results" in data && data.Results != undefined) {
                    this.Results = data.Results;
                }
                if ("HaveErrors" in data && data.HaveErrors != undefined) {
                    this.HaveErrors = data.HaveErrors;
                }
            }
        }
        get BatchID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set BatchID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Results() {
            return pb_1.Message.getRepeatedWrapperField(this, SendQueueMessageResult, 2);
        }
        set Results(value) {
            pb_1.Message.setRepeatedWrapperField(this, 2, value);
        }
        get HaveErrors() {
            return pb_1.Message.getFieldWithDefault(this, 3, false);
        }
        set HaveErrors(value) {
            pb_1.Message.setField(this, 3, value);
        }
        static fromObject(data) {
            const message = new QueueMessagesBatchResponse({});
            if (data.BatchID != null) {
                message.BatchID = data.BatchID;
            }
            if (data.Results != null) {
                message.Results = data.Results.map(item => SendQueueMessageResult.fromObject(item));
            }
            if (data.HaveErrors != null) {
                message.HaveErrors = data.HaveErrors;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.BatchID != null) {
                data.BatchID = this.BatchID;
            }
            if (this.Results != null) {
                data.Results = this.Results.map((item) => item.toObject());
            }
            if (this.HaveErrors != null) {
                data.HaveErrors = this.HaveErrors;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.BatchID.length)
                writer.writeString(1, this.BatchID);
            if (this.Results.length)
                writer.writeRepeatedMessage(2, this.Results, (item) => item.serialize(writer));
            if (this.HaveErrors != false)
                writer.writeBool(3, this.HaveErrors);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueueMessagesBatchResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.BatchID = reader.readString();
                        break;
                    case 2:
                        reader.readMessage(message.Results, () => pb_1.Message.addToRepeatedWrapperField(message, 2, SendQueueMessageResult.deserialize(reader), SendQueueMessageResult));
                        break;
                    case 3:
                        message.HaveErrors = reader.readBool();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueueMessagesBatchResponse.deserialize(bytes);
        }
    }
    _QueueMessagesBatchResponse_one_of_decls = new WeakMap();
    kubemq.QueueMessagesBatchResponse = QueueMessagesBatchResponse;
    class QueueMessageAttributes extends pb_1.Message {
        constructor(data) {
            super();
            _QueueMessageAttributes_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _QueueMessageAttributes_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("Timestamp" in data && data.Timestamp != undefined) {
                    this.Timestamp = data.Timestamp;
                }
                if ("Sequence" in data && data.Sequence != undefined) {
                    this.Sequence = data.Sequence;
                }
                if ("MD5OfBody" in data && data.MD5OfBody != undefined) {
                    this.MD5OfBody = data.MD5OfBody;
                }
                if ("ReceiveCount" in data && data.ReceiveCount != undefined) {
                    this.ReceiveCount = data.ReceiveCount;
                }
                if ("ReRouted" in data && data.ReRouted != undefined) {
                    this.ReRouted = data.ReRouted;
                }
                if ("ReRoutedFromQueue" in data && data.ReRoutedFromQueue != undefined) {
                    this.ReRoutedFromQueue = data.ReRoutedFromQueue;
                }
                if ("ExpirationAt" in data && data.ExpirationAt != undefined) {
                    this.ExpirationAt = data.ExpirationAt;
                }
                if ("DelayedTo" in data && data.DelayedTo != undefined) {
                    this.DelayedTo = data.DelayedTo;
                }
            }
        }
        get Timestamp() {
            return pb_1.Message.getFieldWithDefault(this, 1, 0);
        }
        set Timestamp(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Sequence() {
            return pb_1.Message.getFieldWithDefault(this, 2, 0);
        }
        set Sequence(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get MD5OfBody() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set MD5OfBody(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get ReceiveCount() {
            return pb_1.Message.getFieldWithDefault(this, 4, 0);
        }
        set ReceiveCount(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get ReRouted() {
            return pb_1.Message.getFieldWithDefault(this, 5, false);
        }
        set ReRouted(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get ReRoutedFromQueue() {
            return pb_1.Message.getFieldWithDefault(this, 6, "");
        }
        set ReRoutedFromQueue(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get ExpirationAt() {
            return pb_1.Message.getFieldWithDefault(this, 7, 0);
        }
        set ExpirationAt(value) {
            pb_1.Message.setField(this, 7, value);
        }
        get DelayedTo() {
            return pb_1.Message.getFieldWithDefault(this, 8, 0);
        }
        set DelayedTo(value) {
            pb_1.Message.setField(this, 8, value);
        }
        static fromObject(data) {
            const message = new QueueMessageAttributes({});
            if (data.Timestamp != null) {
                message.Timestamp = data.Timestamp;
            }
            if (data.Sequence != null) {
                message.Sequence = data.Sequence;
            }
            if (data.MD5OfBody != null) {
                message.MD5OfBody = data.MD5OfBody;
            }
            if (data.ReceiveCount != null) {
                message.ReceiveCount = data.ReceiveCount;
            }
            if (data.ReRouted != null) {
                message.ReRouted = data.ReRouted;
            }
            if (data.ReRoutedFromQueue != null) {
                message.ReRoutedFromQueue = data.ReRoutedFromQueue;
            }
            if (data.ExpirationAt != null) {
                message.ExpirationAt = data.ExpirationAt;
            }
            if (data.DelayedTo != null) {
                message.DelayedTo = data.DelayedTo;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.Timestamp != null) {
                data.Timestamp = this.Timestamp;
            }
            if (this.Sequence != null) {
                data.Sequence = this.Sequence;
            }
            if (this.MD5OfBody != null) {
                data.MD5OfBody = this.MD5OfBody;
            }
            if (this.ReceiveCount != null) {
                data.ReceiveCount = this.ReceiveCount;
            }
            if (this.ReRouted != null) {
                data.ReRouted = this.ReRouted;
            }
            if (this.ReRoutedFromQueue != null) {
                data.ReRoutedFromQueue = this.ReRoutedFromQueue;
            }
            if (this.ExpirationAt != null) {
                data.ExpirationAt = this.ExpirationAt;
            }
            if (this.DelayedTo != null) {
                data.DelayedTo = this.DelayedTo;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.Timestamp != 0)
                writer.writeInt64(1, this.Timestamp);
            if (this.Sequence != 0)
                writer.writeUint64(2, this.Sequence);
            if (this.MD5OfBody.length)
                writer.writeString(3, this.MD5OfBody);
            if (this.ReceiveCount != 0)
                writer.writeInt32(4, this.ReceiveCount);
            if (this.ReRouted != false)
                writer.writeBool(5, this.ReRouted);
            if (this.ReRoutedFromQueue.length)
                writer.writeString(6, this.ReRoutedFromQueue);
            if (this.ExpirationAt != 0)
                writer.writeInt64(7, this.ExpirationAt);
            if (this.DelayedTo != 0)
                writer.writeInt64(8, this.DelayedTo);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueueMessageAttributes();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.Timestamp = reader.readInt64();
                        break;
                    case 2:
                        message.Sequence = reader.readUint64();
                        break;
                    case 3:
                        message.MD5OfBody = reader.readString();
                        break;
                    case 4:
                        message.ReceiveCount = reader.readInt32();
                        break;
                    case 5:
                        message.ReRouted = reader.readBool();
                        break;
                    case 6:
                        message.ReRoutedFromQueue = reader.readString();
                        break;
                    case 7:
                        message.ExpirationAt = reader.readInt64();
                        break;
                    case 8:
                        message.DelayedTo = reader.readInt64();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueueMessageAttributes.deserialize(bytes);
        }
    }
    _QueueMessageAttributes_one_of_decls = new WeakMap();
    kubemq.QueueMessageAttributes = QueueMessageAttributes;
    class QueueMessagePolicy extends pb_1.Message {
        constructor(data) {
            super();
            _QueueMessagePolicy_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _QueueMessagePolicy_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("ExpirationSeconds" in data && data.ExpirationSeconds != undefined) {
                    this.ExpirationSeconds = data.ExpirationSeconds;
                }
                if ("DelaySeconds" in data && data.DelaySeconds != undefined) {
                    this.DelaySeconds = data.DelaySeconds;
                }
                if ("MaxReceiveCount" in data && data.MaxReceiveCount != undefined) {
                    this.MaxReceiveCount = data.MaxReceiveCount;
                }
                if ("MaxReceiveQueue" in data && data.MaxReceiveQueue != undefined) {
                    this.MaxReceiveQueue = data.MaxReceiveQueue;
                }
            }
        }
        get ExpirationSeconds() {
            return pb_1.Message.getFieldWithDefault(this, 1, 0);
        }
        set ExpirationSeconds(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get DelaySeconds() {
            return pb_1.Message.getFieldWithDefault(this, 2, 0);
        }
        set DelaySeconds(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get MaxReceiveCount() {
            return pb_1.Message.getFieldWithDefault(this, 3, 0);
        }
        set MaxReceiveCount(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get MaxReceiveQueue() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set MaxReceiveQueue(value) {
            pb_1.Message.setField(this, 4, value);
        }
        static fromObject(data) {
            const message = new QueueMessagePolicy({});
            if (data.ExpirationSeconds != null) {
                message.ExpirationSeconds = data.ExpirationSeconds;
            }
            if (data.DelaySeconds != null) {
                message.DelaySeconds = data.DelaySeconds;
            }
            if (data.MaxReceiveCount != null) {
                message.MaxReceiveCount = data.MaxReceiveCount;
            }
            if (data.MaxReceiveQueue != null) {
                message.MaxReceiveQueue = data.MaxReceiveQueue;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.ExpirationSeconds != null) {
                data.ExpirationSeconds = this.ExpirationSeconds;
            }
            if (this.DelaySeconds != null) {
                data.DelaySeconds = this.DelaySeconds;
            }
            if (this.MaxReceiveCount != null) {
                data.MaxReceiveCount = this.MaxReceiveCount;
            }
            if (this.MaxReceiveQueue != null) {
                data.MaxReceiveQueue = this.MaxReceiveQueue;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.ExpirationSeconds != 0)
                writer.writeInt32(1, this.ExpirationSeconds);
            if (this.DelaySeconds != 0)
                writer.writeInt32(2, this.DelaySeconds);
            if (this.MaxReceiveCount != 0)
                writer.writeInt32(3, this.MaxReceiveCount);
            if (this.MaxReceiveQueue.length)
                writer.writeString(4, this.MaxReceiveQueue);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueueMessagePolicy();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.ExpirationSeconds = reader.readInt32();
                        break;
                    case 2:
                        message.DelaySeconds = reader.readInt32();
                        break;
                    case 3:
                        message.MaxReceiveCount = reader.readInt32();
                        break;
                    case 4:
                        message.MaxReceiveQueue = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueueMessagePolicy.deserialize(bytes);
        }
    }
    _QueueMessagePolicy_one_of_decls = new WeakMap();
    kubemq.QueueMessagePolicy = QueueMessagePolicy;
    class SendQueueMessageResult extends pb_1.Message {
        constructor(data) {
            super();
            _SendQueueMessageResult_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _SendQueueMessageResult_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("MessageID" in data && data.MessageID != undefined) {
                    this.MessageID = data.MessageID;
                }
                if ("SentAt" in data && data.SentAt != undefined) {
                    this.SentAt = data.SentAt;
                }
                if ("ExpirationAt" in data && data.ExpirationAt != undefined) {
                    this.ExpirationAt = data.ExpirationAt;
                }
                if ("DelayedTo" in data && data.DelayedTo != undefined) {
                    this.DelayedTo = data.DelayedTo;
                }
                if ("IsError" in data && data.IsError != undefined) {
                    this.IsError = data.IsError;
                }
                if ("Error" in data && data.Error != undefined) {
                    this.Error = data.Error;
                }
                if ("RefChannel" in data && data.RefChannel != undefined) {
                    this.RefChannel = data.RefChannel;
                }
                if ("RefTopic" in data && data.RefTopic != undefined) {
                    this.RefTopic = data.RefTopic;
                }
                if ("RefPartition" in data && data.RefPartition != undefined) {
                    this.RefPartition = data.RefPartition;
                }
                if ("RefHash" in data && data.RefHash != undefined) {
                    this.RefHash = data.RefHash;
                }
            }
        }
        get MessageID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set MessageID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get SentAt() {
            return pb_1.Message.getFieldWithDefault(this, 2, 0);
        }
        set SentAt(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get ExpirationAt() {
            return pb_1.Message.getFieldWithDefault(this, 3, 0);
        }
        set ExpirationAt(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get DelayedTo() {
            return pb_1.Message.getFieldWithDefault(this, 4, 0);
        }
        set DelayedTo(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get IsError() {
            return pb_1.Message.getFieldWithDefault(this, 5, false);
        }
        set IsError(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get Error() {
            return pb_1.Message.getFieldWithDefault(this, 6, "");
        }
        set Error(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get RefChannel() {
            return pb_1.Message.getFieldWithDefault(this, 7, "");
        }
        set RefChannel(value) {
            pb_1.Message.setField(this, 7, value);
        }
        get RefTopic() {
            return pb_1.Message.getFieldWithDefault(this, 8, "");
        }
        set RefTopic(value) {
            pb_1.Message.setField(this, 8, value);
        }
        get RefPartition() {
            return pb_1.Message.getFieldWithDefault(this, 9, 0);
        }
        set RefPartition(value) {
            pb_1.Message.setField(this, 9, value);
        }
        get RefHash() {
            return pb_1.Message.getFieldWithDefault(this, 10, "");
        }
        set RefHash(value) {
            pb_1.Message.setField(this, 10, value);
        }
        static fromObject(data) {
            const message = new SendQueueMessageResult({});
            if (data.MessageID != null) {
                message.MessageID = data.MessageID;
            }
            if (data.SentAt != null) {
                message.SentAt = data.SentAt;
            }
            if (data.ExpirationAt != null) {
                message.ExpirationAt = data.ExpirationAt;
            }
            if (data.DelayedTo != null) {
                message.DelayedTo = data.DelayedTo;
            }
            if (data.IsError != null) {
                message.IsError = data.IsError;
            }
            if (data.Error != null) {
                message.Error = data.Error;
            }
            if (data.RefChannel != null) {
                message.RefChannel = data.RefChannel;
            }
            if (data.RefTopic != null) {
                message.RefTopic = data.RefTopic;
            }
            if (data.RefPartition != null) {
                message.RefPartition = data.RefPartition;
            }
            if (data.RefHash != null) {
                message.RefHash = data.RefHash;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.MessageID != null) {
                data.MessageID = this.MessageID;
            }
            if (this.SentAt != null) {
                data.SentAt = this.SentAt;
            }
            if (this.ExpirationAt != null) {
                data.ExpirationAt = this.ExpirationAt;
            }
            if (this.DelayedTo != null) {
                data.DelayedTo = this.DelayedTo;
            }
            if (this.IsError != null) {
                data.IsError = this.IsError;
            }
            if (this.Error != null) {
                data.Error = this.Error;
            }
            if (this.RefChannel != null) {
                data.RefChannel = this.RefChannel;
            }
            if (this.RefTopic != null) {
                data.RefTopic = this.RefTopic;
            }
            if (this.RefPartition != null) {
                data.RefPartition = this.RefPartition;
            }
            if (this.RefHash != null) {
                data.RefHash = this.RefHash;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.MessageID.length)
                writer.writeString(1, this.MessageID);
            if (this.SentAt != 0)
                writer.writeInt64(2, this.SentAt);
            if (this.ExpirationAt != 0)
                writer.writeInt64(3, this.ExpirationAt);
            if (this.DelayedTo != 0)
                writer.writeInt64(4, this.DelayedTo);
            if (this.IsError != false)
                writer.writeBool(5, this.IsError);
            if (this.Error.length)
                writer.writeString(6, this.Error);
            if (this.RefChannel.length)
                writer.writeString(7, this.RefChannel);
            if (this.RefTopic.length)
                writer.writeString(8, this.RefTopic);
            if (this.RefPartition != 0)
                writer.writeInt32(9, this.RefPartition);
            if (this.RefHash.length)
                writer.writeString(10, this.RefHash);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new SendQueueMessageResult();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.MessageID = reader.readString();
                        break;
                    case 2:
                        message.SentAt = reader.readInt64();
                        break;
                    case 3:
                        message.ExpirationAt = reader.readInt64();
                        break;
                    case 4:
                        message.DelayedTo = reader.readInt64();
                        break;
                    case 5:
                        message.IsError = reader.readBool();
                        break;
                    case 6:
                        message.Error = reader.readString();
                        break;
                    case 7:
                        message.RefChannel = reader.readString();
                        break;
                    case 8:
                        message.RefTopic = reader.readString();
                        break;
                    case 9:
                        message.RefPartition = reader.readInt32();
                        break;
                    case 10:
                        message.RefHash = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return SendQueueMessageResult.deserialize(bytes);
        }
    }
    _SendQueueMessageResult_one_of_decls = new WeakMap();
    kubemq.SendQueueMessageResult = SendQueueMessageResult;
    class ReceiveQueueMessagesRequest extends pb_1.Message {
        constructor(data) {
            super();
            _ReceiveQueueMessagesRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _ReceiveQueueMessagesRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("ClientID" in data && data.ClientID != undefined) {
                    this.ClientID = data.ClientID;
                }
                if ("Channel" in data && data.Channel != undefined) {
                    this.Channel = data.Channel;
                }
                if ("MaxNumberOfMessages" in data && data.MaxNumberOfMessages != undefined) {
                    this.MaxNumberOfMessages = data.MaxNumberOfMessages;
                }
                if ("WaitTimeSeconds" in data && data.WaitTimeSeconds != undefined) {
                    this.WaitTimeSeconds = data.WaitTimeSeconds;
                }
                if ("IsPeak" in data && data.IsPeak != undefined) {
                    this.IsPeak = data.IsPeak;
                }
            }
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get ClientID() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set ClientID(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Channel() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set Channel(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get MaxNumberOfMessages() {
            return pb_1.Message.getFieldWithDefault(this, 4, 0);
        }
        set MaxNumberOfMessages(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get WaitTimeSeconds() {
            return pb_1.Message.getFieldWithDefault(this, 5, 0);
        }
        set WaitTimeSeconds(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get IsPeak() {
            return pb_1.Message.getFieldWithDefault(this, 6, false);
        }
        set IsPeak(value) {
            pb_1.Message.setField(this, 6, value);
        }
        static fromObject(data) {
            const message = new ReceiveQueueMessagesRequest({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.ClientID != null) {
                message.ClientID = data.ClientID;
            }
            if (data.Channel != null) {
                message.Channel = data.Channel;
            }
            if (data.MaxNumberOfMessages != null) {
                message.MaxNumberOfMessages = data.MaxNumberOfMessages;
            }
            if (data.WaitTimeSeconds != null) {
                message.WaitTimeSeconds = data.WaitTimeSeconds;
            }
            if (data.IsPeak != null) {
                message.IsPeak = data.IsPeak;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.ClientID != null) {
                data.ClientID = this.ClientID;
            }
            if (this.Channel != null) {
                data.Channel = this.Channel;
            }
            if (this.MaxNumberOfMessages != null) {
                data.MaxNumberOfMessages = this.MaxNumberOfMessages;
            }
            if (this.WaitTimeSeconds != null) {
                data.WaitTimeSeconds = this.WaitTimeSeconds;
            }
            if (this.IsPeak != null) {
                data.IsPeak = this.IsPeak;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.ClientID.length)
                writer.writeString(2, this.ClientID);
            if (this.Channel.length)
                writer.writeString(3, this.Channel);
            if (this.MaxNumberOfMessages != 0)
                writer.writeInt32(4, this.MaxNumberOfMessages);
            if (this.WaitTimeSeconds != 0)
                writer.writeInt32(5, this.WaitTimeSeconds);
            if (this.IsPeak != false)
                writer.writeBool(6, this.IsPeak);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new ReceiveQueueMessagesRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        message.ClientID = reader.readString();
                        break;
                    case 3:
                        message.Channel = reader.readString();
                        break;
                    case 4:
                        message.MaxNumberOfMessages = reader.readInt32();
                        break;
                    case 5:
                        message.WaitTimeSeconds = reader.readInt32();
                        break;
                    case 6:
                        message.IsPeak = reader.readBool();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return ReceiveQueueMessagesRequest.deserialize(bytes);
        }
    }
    _ReceiveQueueMessagesRequest_one_of_decls = new WeakMap();
    kubemq.ReceiveQueueMessagesRequest = ReceiveQueueMessagesRequest;
    class ReceiveQueueMessagesResponse extends pb_1.Message {
        constructor(data) {
            super();
            _ReceiveQueueMessagesResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [2], tslib_1.__classPrivateFieldGet(this, _ReceiveQueueMessagesResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("Messages" in data && data.Messages != undefined) {
                    this.Messages = data.Messages;
                }
                if ("MessagesReceived" in data && data.MessagesReceived != undefined) {
                    this.MessagesReceived = data.MessagesReceived;
                }
                if ("MessagesExpired" in data && data.MessagesExpired != undefined) {
                    this.MessagesExpired = data.MessagesExpired;
                }
                if ("IsPeak" in data && data.IsPeak != undefined) {
                    this.IsPeak = data.IsPeak;
                }
                if ("IsError" in data && data.IsError != undefined) {
                    this.IsError = data.IsError;
                }
                if ("Error" in data && data.Error != undefined) {
                    this.Error = data.Error;
                }
            }
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Messages() {
            return pb_1.Message.getRepeatedWrapperField(this, QueueMessage, 2);
        }
        set Messages(value) {
            pb_1.Message.setRepeatedWrapperField(this, 2, value);
        }
        get MessagesReceived() {
            return pb_1.Message.getFieldWithDefault(this, 3, 0);
        }
        set MessagesReceived(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get MessagesExpired() {
            return pb_1.Message.getFieldWithDefault(this, 4, 0);
        }
        set MessagesExpired(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get IsPeak() {
            return pb_1.Message.getFieldWithDefault(this, 5, false);
        }
        set IsPeak(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get IsError() {
            return pb_1.Message.getFieldWithDefault(this, 6, false);
        }
        set IsError(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get Error() {
            return pb_1.Message.getFieldWithDefault(this, 7, "");
        }
        set Error(value) {
            pb_1.Message.setField(this, 7, value);
        }
        static fromObject(data) {
            const message = new ReceiveQueueMessagesResponse({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.Messages != null) {
                message.Messages = data.Messages.map(item => QueueMessage.fromObject(item));
            }
            if (data.MessagesReceived != null) {
                message.MessagesReceived = data.MessagesReceived;
            }
            if (data.MessagesExpired != null) {
                message.MessagesExpired = data.MessagesExpired;
            }
            if (data.IsPeak != null) {
                message.IsPeak = data.IsPeak;
            }
            if (data.IsError != null) {
                message.IsError = data.IsError;
            }
            if (data.Error != null) {
                message.Error = data.Error;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.Messages != null) {
                data.Messages = this.Messages.map((item) => item.toObject());
            }
            if (this.MessagesReceived != null) {
                data.MessagesReceived = this.MessagesReceived;
            }
            if (this.MessagesExpired != null) {
                data.MessagesExpired = this.MessagesExpired;
            }
            if (this.IsPeak != null) {
                data.IsPeak = this.IsPeak;
            }
            if (this.IsError != null) {
                data.IsError = this.IsError;
            }
            if (this.Error != null) {
                data.Error = this.Error;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.Messages.length)
                writer.writeRepeatedMessage(2, this.Messages, (item) => item.serialize(writer));
            if (this.MessagesReceived != 0)
                writer.writeInt32(3, this.MessagesReceived);
            if (this.MessagesExpired != 0)
                writer.writeInt32(4, this.MessagesExpired);
            if (this.IsPeak != false)
                writer.writeBool(5, this.IsPeak);
            if (this.IsError != false)
                writer.writeBool(6, this.IsError);
            if (this.Error.length)
                writer.writeString(7, this.Error);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new ReceiveQueueMessagesResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        reader.readMessage(message.Messages, () => pb_1.Message.addToRepeatedWrapperField(message, 2, QueueMessage.deserialize(reader), QueueMessage));
                        break;
                    case 3:
                        message.MessagesReceived = reader.readInt32();
                        break;
                    case 4:
                        message.MessagesExpired = reader.readInt32();
                        break;
                    case 5:
                        message.IsPeak = reader.readBool();
                        break;
                    case 6:
                        message.IsError = reader.readBool();
                        break;
                    case 7:
                        message.Error = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return ReceiveQueueMessagesResponse.deserialize(bytes);
        }
    }
    _ReceiveQueueMessagesResponse_one_of_decls = new WeakMap();
    kubemq.ReceiveQueueMessagesResponse = ReceiveQueueMessagesResponse;
    class AckAllQueueMessagesRequest extends pb_1.Message {
        constructor(data) {
            super();
            _AckAllQueueMessagesRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _AckAllQueueMessagesRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("ClientID" in data && data.ClientID != undefined) {
                    this.ClientID = data.ClientID;
                }
                if ("Channel" in data && data.Channel != undefined) {
                    this.Channel = data.Channel;
                }
                if ("WaitTimeSeconds" in data && data.WaitTimeSeconds != undefined) {
                    this.WaitTimeSeconds = data.WaitTimeSeconds;
                }
            }
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get ClientID() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set ClientID(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Channel() {
            return pb_1.Message.getFieldWithDefault(this, 3, "");
        }
        set Channel(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get WaitTimeSeconds() {
            return pb_1.Message.getFieldWithDefault(this, 4, 0);
        }
        set WaitTimeSeconds(value) {
            pb_1.Message.setField(this, 4, value);
        }
        static fromObject(data) {
            const message = new AckAllQueueMessagesRequest({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.ClientID != null) {
                message.ClientID = data.ClientID;
            }
            if (data.Channel != null) {
                message.Channel = data.Channel;
            }
            if (data.WaitTimeSeconds != null) {
                message.WaitTimeSeconds = data.WaitTimeSeconds;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.ClientID != null) {
                data.ClientID = this.ClientID;
            }
            if (this.Channel != null) {
                data.Channel = this.Channel;
            }
            if (this.WaitTimeSeconds != null) {
                data.WaitTimeSeconds = this.WaitTimeSeconds;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.ClientID.length)
                writer.writeString(2, this.ClientID);
            if (this.Channel.length)
                writer.writeString(3, this.Channel);
            if (this.WaitTimeSeconds != 0)
                writer.writeInt32(4, this.WaitTimeSeconds);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new AckAllQueueMessagesRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        message.ClientID = reader.readString();
                        break;
                    case 3:
                        message.Channel = reader.readString();
                        break;
                    case 4:
                        message.WaitTimeSeconds = reader.readInt32();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return AckAllQueueMessagesRequest.deserialize(bytes);
        }
    }
    _AckAllQueueMessagesRequest_one_of_decls = new WeakMap();
    kubemq.AckAllQueueMessagesRequest = AckAllQueueMessagesRequest;
    class AckAllQueueMessagesResponse extends pb_1.Message {
        constructor(data) {
            super();
            _AckAllQueueMessagesResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _AckAllQueueMessagesResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("AffectedMessages" in data && data.AffectedMessages != undefined) {
                    this.AffectedMessages = data.AffectedMessages;
                }
                if ("IsError" in data && data.IsError != undefined) {
                    this.IsError = data.IsError;
                }
                if ("Error" in data && data.Error != undefined) {
                    this.Error = data.Error;
                }
            }
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get AffectedMessages() {
            return pb_1.Message.getFieldWithDefault(this, 2, 0);
        }
        set AffectedMessages(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get IsError() {
            return pb_1.Message.getFieldWithDefault(this, 3, false);
        }
        set IsError(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Error() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set Error(value) {
            pb_1.Message.setField(this, 4, value);
        }
        static fromObject(data) {
            const message = new AckAllQueueMessagesResponse({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.AffectedMessages != null) {
                message.AffectedMessages = data.AffectedMessages;
            }
            if (data.IsError != null) {
                message.IsError = data.IsError;
            }
            if (data.Error != null) {
                message.Error = data.Error;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.AffectedMessages != null) {
                data.AffectedMessages = this.AffectedMessages;
            }
            if (this.IsError != null) {
                data.IsError = this.IsError;
            }
            if (this.Error != null) {
                data.Error = this.Error;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.AffectedMessages != 0)
                writer.writeUint64(2, this.AffectedMessages);
            if (this.IsError != false)
                writer.writeBool(3, this.IsError);
            if (this.Error.length)
                writer.writeString(4, this.Error);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new AckAllQueueMessagesResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        message.AffectedMessages = reader.readUint64();
                        break;
                    case 3:
                        message.IsError = reader.readBool();
                        break;
                    case 4:
                        message.Error = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return AckAllQueueMessagesResponse.deserialize(bytes);
        }
    }
    _AckAllQueueMessagesResponse_one_of_decls = new WeakMap();
    kubemq.AckAllQueueMessagesResponse = AckAllQueueMessagesResponse;
    class StreamQueueMessagesRequest extends pb_1.Message {
        constructor(data) {
            super();
            _StreamQueueMessagesRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _StreamQueueMessagesRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("ClientID" in data && data.ClientID != undefined) {
                    this.ClientID = data.ClientID;
                }
                if ("StreamRequestTypeData" in data && data.StreamRequestTypeData != undefined) {
                    this.StreamRequestTypeData = data.StreamRequestTypeData;
                }
                if ("Channel" in data && data.Channel != undefined) {
                    this.Channel = data.Channel;
                }
                if ("VisibilitySeconds" in data && data.VisibilitySeconds != undefined) {
                    this.VisibilitySeconds = data.VisibilitySeconds;
                }
                if ("WaitTimeSeconds" in data && data.WaitTimeSeconds != undefined) {
                    this.WaitTimeSeconds = data.WaitTimeSeconds;
                }
                if ("RefSequence" in data && data.RefSequence != undefined) {
                    this.RefSequence = data.RefSequence;
                }
                if ("ModifiedMessage" in data && data.ModifiedMessage != undefined) {
                    this.ModifiedMessage = data.ModifiedMessage;
                }
            }
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get ClientID() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set ClientID(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get StreamRequestTypeData() {
            return pb_1.Message.getFieldWithDefault(this, 3, StreamRequestType.StreamRequestTypeUnknown);
        }
        set StreamRequestTypeData(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Channel() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set Channel(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get VisibilitySeconds() {
            return pb_1.Message.getFieldWithDefault(this, 5, 0);
        }
        set VisibilitySeconds(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get WaitTimeSeconds() {
            return pb_1.Message.getFieldWithDefault(this, 6, 0);
        }
        set WaitTimeSeconds(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get RefSequence() {
            return pb_1.Message.getFieldWithDefault(this, 7, 0);
        }
        set RefSequence(value) {
            pb_1.Message.setField(this, 7, value);
        }
        get ModifiedMessage() {
            return pb_1.Message.getWrapperField(this, QueueMessage, 8);
        }
        set ModifiedMessage(value) {
            pb_1.Message.setWrapperField(this, 8, value);
        }
        get has_ModifiedMessage() {
            return pb_1.Message.getField(this, 8) != null;
        }
        static fromObject(data) {
            const message = new StreamQueueMessagesRequest({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.ClientID != null) {
                message.ClientID = data.ClientID;
            }
            if (data.StreamRequestTypeData != null) {
                message.StreamRequestTypeData = data.StreamRequestTypeData;
            }
            if (data.Channel != null) {
                message.Channel = data.Channel;
            }
            if (data.VisibilitySeconds != null) {
                message.VisibilitySeconds = data.VisibilitySeconds;
            }
            if (data.WaitTimeSeconds != null) {
                message.WaitTimeSeconds = data.WaitTimeSeconds;
            }
            if (data.RefSequence != null) {
                message.RefSequence = data.RefSequence;
            }
            if (data.ModifiedMessage != null) {
                message.ModifiedMessage = QueueMessage.fromObject(data.ModifiedMessage);
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.ClientID != null) {
                data.ClientID = this.ClientID;
            }
            if (this.StreamRequestTypeData != null) {
                data.StreamRequestTypeData = this.StreamRequestTypeData;
            }
            if (this.Channel != null) {
                data.Channel = this.Channel;
            }
            if (this.VisibilitySeconds != null) {
                data.VisibilitySeconds = this.VisibilitySeconds;
            }
            if (this.WaitTimeSeconds != null) {
                data.WaitTimeSeconds = this.WaitTimeSeconds;
            }
            if (this.RefSequence != null) {
                data.RefSequence = this.RefSequence;
            }
            if (this.ModifiedMessage != null) {
                data.ModifiedMessage = this.ModifiedMessage.toObject();
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.ClientID.length)
                writer.writeString(2, this.ClientID);
            if (this.StreamRequestTypeData != StreamRequestType.StreamRequestTypeUnknown)
                writer.writeEnum(3, this.StreamRequestTypeData);
            if (this.Channel.length)
                writer.writeString(4, this.Channel);
            if (this.VisibilitySeconds != 0)
                writer.writeInt32(5, this.VisibilitySeconds);
            if (this.WaitTimeSeconds != 0)
                writer.writeInt32(6, this.WaitTimeSeconds);
            if (this.RefSequence != 0)
                writer.writeUint64(7, this.RefSequence);
            if (this.has_ModifiedMessage)
                writer.writeMessage(8, this.ModifiedMessage, () => this.ModifiedMessage.serialize(writer));
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new StreamQueueMessagesRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        message.ClientID = reader.readString();
                        break;
                    case 3:
                        message.StreamRequestTypeData = reader.readEnum();
                        break;
                    case 4:
                        message.Channel = reader.readString();
                        break;
                    case 5:
                        message.VisibilitySeconds = reader.readInt32();
                        break;
                    case 6:
                        message.WaitTimeSeconds = reader.readInt32();
                        break;
                    case 7:
                        message.RefSequence = reader.readUint64();
                        break;
                    case 8:
                        reader.readMessage(message.ModifiedMessage, () => message.ModifiedMessage = QueueMessage.deserialize(reader));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return StreamQueueMessagesRequest.deserialize(bytes);
        }
    }
    _StreamQueueMessagesRequest_one_of_decls = new WeakMap();
    kubemq.StreamQueueMessagesRequest = StreamQueueMessagesRequest;
    class StreamQueueMessagesResponse extends pb_1.Message {
        constructor(data) {
            super();
            _StreamQueueMessagesResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _StreamQueueMessagesResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("StreamRequestTypeData" in data && data.StreamRequestTypeData != undefined) {
                    this.StreamRequestTypeData = data.StreamRequestTypeData;
                }
                if ("Message" in data && data.Message != undefined) {
                    this.Message = data.Message;
                }
                if ("IsError" in data && data.IsError != undefined) {
                    this.IsError = data.IsError;
                }
                if ("Error" in data && data.Error != undefined) {
                    this.Error = data.Error;
                }
            }
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get StreamRequestTypeData() {
            return pb_1.Message.getFieldWithDefault(this, 2, StreamRequestType.StreamRequestTypeUnknown);
        }
        set StreamRequestTypeData(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Message() {
            return pb_1.Message.getWrapperField(this, QueueMessage, 3);
        }
        set Message(value) {
            pb_1.Message.setWrapperField(this, 3, value);
        }
        get has_Message() {
            return pb_1.Message.getField(this, 3) != null;
        }
        get IsError() {
            return pb_1.Message.getFieldWithDefault(this, 4, false);
        }
        set IsError(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get Error() {
            return pb_1.Message.getFieldWithDefault(this, 5, "");
        }
        set Error(value) {
            pb_1.Message.setField(this, 5, value);
        }
        static fromObject(data) {
            const message = new StreamQueueMessagesResponse({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.StreamRequestTypeData != null) {
                message.StreamRequestTypeData = data.StreamRequestTypeData;
            }
            if (data.Message != null) {
                message.Message = QueueMessage.fromObject(data.Message);
            }
            if (data.IsError != null) {
                message.IsError = data.IsError;
            }
            if (data.Error != null) {
                message.Error = data.Error;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.StreamRequestTypeData != null) {
                data.StreamRequestTypeData = this.StreamRequestTypeData;
            }
            if (this.Message != null) {
                data.Message = this.Message.toObject();
            }
            if (this.IsError != null) {
                data.IsError = this.IsError;
            }
            if (this.Error != null) {
                data.Error = this.Error;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.StreamRequestTypeData != StreamRequestType.StreamRequestTypeUnknown)
                writer.writeEnum(2, this.StreamRequestTypeData);
            if (this.has_Message)
                writer.writeMessage(3, this.Message, () => this.Message.serialize(writer));
            if (this.IsError != false)
                writer.writeBool(4, this.IsError);
            if (this.Error.length)
                writer.writeString(5, this.Error);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new StreamQueueMessagesResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        message.StreamRequestTypeData = reader.readEnum();
                        break;
                    case 3:
                        reader.readMessage(message.Message, () => message.Message = QueueMessage.deserialize(reader));
                        break;
                    case 4:
                        message.IsError = reader.readBool();
                        break;
                    case 5:
                        message.Error = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return StreamQueueMessagesResponse.deserialize(bytes);
        }
    }
    _StreamQueueMessagesResponse_one_of_decls = new WeakMap();
    kubemq.StreamQueueMessagesResponse = StreamQueueMessagesResponse;
    class QueuesUpstreamRequest extends pb_1.Message {
        constructor(data) {
            super();
            _QueuesUpstreamRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [2], tslib_1.__classPrivateFieldGet(this, _QueuesUpstreamRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("Messages" in data && data.Messages != undefined) {
                    this.Messages = data.Messages;
                }
            }
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Messages() {
            return pb_1.Message.getRepeatedWrapperField(this, QueueMessage, 2);
        }
        set Messages(value) {
            pb_1.Message.setRepeatedWrapperField(this, 2, value);
        }
        static fromObject(data) {
            const message = new QueuesUpstreamRequest({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.Messages != null) {
                message.Messages = data.Messages.map(item => QueueMessage.fromObject(item));
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.Messages != null) {
                data.Messages = this.Messages.map((item) => item.toObject());
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.Messages.length)
                writer.writeRepeatedMessage(2, this.Messages, (item) => item.serialize(writer));
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueuesUpstreamRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        reader.readMessage(message.Messages, () => pb_1.Message.addToRepeatedWrapperField(message, 2, QueueMessage.deserialize(reader), QueueMessage));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueuesUpstreamRequest.deserialize(bytes);
        }
    }
    _QueuesUpstreamRequest_one_of_decls = new WeakMap();
    kubemq.QueuesUpstreamRequest = QueuesUpstreamRequest;
    class QueuesUpstreamResponse extends pb_1.Message {
        constructor(data) {
            super();
            _QueuesUpstreamResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [2], tslib_1.__classPrivateFieldGet(this, _QueuesUpstreamResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RefRequestID" in data && data.RefRequestID != undefined) {
                    this.RefRequestID = data.RefRequestID;
                }
                if ("Results" in data && data.Results != undefined) {
                    this.Results = data.Results;
                }
                if ("IsError" in data && data.IsError != undefined) {
                    this.IsError = data.IsError;
                }
                if ("Error" in data && data.Error != undefined) {
                    this.Error = data.Error;
                }
            }
        }
        get RefRequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RefRequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Results() {
            return pb_1.Message.getRepeatedWrapperField(this, SendQueueMessageResult, 2);
        }
        set Results(value) {
            pb_1.Message.setRepeatedWrapperField(this, 2, value);
        }
        get IsError() {
            return pb_1.Message.getFieldWithDefault(this, 3, false);
        }
        set IsError(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Error() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set Error(value) {
            pb_1.Message.setField(this, 4, value);
        }
        static fromObject(data) {
            const message = new QueuesUpstreamResponse({});
            if (data.RefRequestID != null) {
                message.RefRequestID = data.RefRequestID;
            }
            if (data.Results != null) {
                message.Results = data.Results.map(item => SendQueueMessageResult.fromObject(item));
            }
            if (data.IsError != null) {
                message.IsError = data.IsError;
            }
            if (data.Error != null) {
                message.Error = data.Error;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RefRequestID != null) {
                data.RefRequestID = this.RefRequestID;
            }
            if (this.Results != null) {
                data.Results = this.Results.map((item) => item.toObject());
            }
            if (this.IsError != null) {
                data.IsError = this.IsError;
            }
            if (this.Error != null) {
                data.Error = this.Error;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RefRequestID.length)
                writer.writeString(1, this.RefRequestID);
            if (this.Results.length)
                writer.writeRepeatedMessage(2, this.Results, (item) => item.serialize(writer));
            if (this.IsError != false)
                writer.writeBool(3, this.IsError);
            if (this.Error.length)
                writer.writeString(4, this.Error);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueuesUpstreamResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RefRequestID = reader.readString();
                        break;
                    case 2:
                        reader.readMessage(message.Results, () => pb_1.Message.addToRepeatedWrapperField(message, 2, SendQueueMessageResult.deserialize(reader), SendQueueMessageResult));
                        break;
                    case 3:
                        message.IsError = reader.readBool();
                        break;
                    case 4:
                        message.Error = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueuesUpstreamResponse.deserialize(bytes);
        }
    }
    _QueuesUpstreamResponse_one_of_decls = new WeakMap();
    kubemq.QueuesUpstreamResponse = QueuesUpstreamResponse;
    class QueuesDownstreamRequest extends pb_1.Message {
        constructor(data) {
            super();
            _QueuesDownstreamRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [9], tslib_1.__classPrivateFieldGet(this, _QueuesDownstreamRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("ClientID" in data && data.ClientID != undefined) {
                    this.ClientID = data.ClientID;
                }
                if ("RequestTypeData" in data && data.RequestTypeData != undefined) {
                    this.RequestTypeData = data.RequestTypeData;
                }
                if ("Channel" in data && data.Channel != undefined) {
                    this.Channel = data.Channel;
                }
                if ("MaxItems" in data && data.MaxItems != undefined) {
                    this.MaxItems = data.MaxItems;
                }
                if ("WaitTimeout" in data && data.WaitTimeout != undefined) {
                    this.WaitTimeout = data.WaitTimeout;
                }
                if ("AutoAck" in data && data.AutoAck != undefined) {
                    this.AutoAck = data.AutoAck;
                }
                if ("ReQueueChannel" in data && data.ReQueueChannel != undefined) {
                    this.ReQueueChannel = data.ReQueueChannel;
                }
                if ("SequenceRange" in data && data.SequenceRange != undefined) {
                    this.SequenceRange = data.SequenceRange;
                }
                if ("RefTransactionId" in data && data.RefTransactionId != undefined) {
                    this.RefTransactionId = data.RefTransactionId;
                }
                if ("Metadata" in data && data.Metadata != undefined) {
                    this.Metadata = data.Metadata;
                }
            }
            if (!this.Metadata)
                this.Metadata = new Map();
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get ClientID() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set ClientID(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get RequestTypeData() {
            return pb_1.Message.getFieldWithDefault(this, 3, QueuesDownstreamRequestType.PollRequestTypeUnknown);
        }
        set RequestTypeData(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Channel() {
            return pb_1.Message.getFieldWithDefault(this, 4, "");
        }
        set Channel(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get MaxItems() {
            return pb_1.Message.getFieldWithDefault(this, 5, 0);
        }
        set MaxItems(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get WaitTimeout() {
            return pb_1.Message.getFieldWithDefault(this, 6, 0);
        }
        set WaitTimeout(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get AutoAck() {
            return pb_1.Message.getFieldWithDefault(this, 7, false);
        }
        set AutoAck(value) {
            pb_1.Message.setField(this, 7, value);
        }
        get ReQueueChannel() {
            return pb_1.Message.getFieldWithDefault(this, 8, "");
        }
        set ReQueueChannel(value) {
            pb_1.Message.setField(this, 8, value);
        }
        get SequenceRange() {
            return pb_1.Message.getFieldWithDefault(this, 9, []);
        }
        set SequenceRange(value) {
            pb_1.Message.setField(this, 9, value);
        }
        get RefTransactionId() {
            return pb_1.Message.getFieldWithDefault(this, 10, "");
        }
        set RefTransactionId(value) {
            pb_1.Message.setField(this, 10, value);
        }
        get Metadata() {
            return pb_1.Message.getField(this, 12);
        }
        set Metadata(value) {
            pb_1.Message.setField(this, 12, value);
        }
        static fromObject(data) {
            const message = new QueuesDownstreamRequest({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.ClientID != null) {
                message.ClientID = data.ClientID;
            }
            if (data.RequestTypeData != null) {
                message.RequestTypeData = data.RequestTypeData;
            }
            if (data.Channel != null) {
                message.Channel = data.Channel;
            }
            if (data.MaxItems != null) {
                message.MaxItems = data.MaxItems;
            }
            if (data.WaitTimeout != null) {
                message.WaitTimeout = data.WaitTimeout;
            }
            if (data.AutoAck != null) {
                message.AutoAck = data.AutoAck;
            }
            if (data.ReQueueChannel != null) {
                message.ReQueueChannel = data.ReQueueChannel;
            }
            if (data.SequenceRange != null) {
                message.SequenceRange = data.SequenceRange;
            }
            if (data.RefTransactionId != null) {
                message.RefTransactionId = data.RefTransactionId;
            }
            if (typeof data.Metadata == "object") {
                message.Metadata = new Map(Object.entries(data.Metadata));
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.ClientID != null) {
                data.ClientID = this.ClientID;
            }
            if (this.RequestTypeData != null) {
                data.RequestTypeData = this.RequestTypeData;
            }
            if (this.Channel != null) {
                data.Channel = this.Channel;
            }
            if (this.MaxItems != null) {
                data.MaxItems = this.MaxItems;
            }
            if (this.WaitTimeout != null) {
                data.WaitTimeout = this.WaitTimeout;
            }
            if (this.AutoAck != null) {
                data.AutoAck = this.AutoAck;
            }
            if (this.ReQueueChannel != null) {
                data.ReQueueChannel = this.ReQueueChannel;
            }
            if (this.SequenceRange != null) {
                data.SequenceRange = this.SequenceRange;
            }
            if (this.RefTransactionId != null) {
                data.RefTransactionId = this.RefTransactionId;
            }
            if (this.Metadata != null) {
                data.Metadata = (Object.fromEntries)(this.Metadata);
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.ClientID.length)
                writer.writeString(2, this.ClientID);
            if (this.RequestTypeData != QueuesDownstreamRequestType.PollRequestTypeUnknown)
                writer.writeEnum(3, this.RequestTypeData);
            if (this.Channel.length)
                writer.writeString(4, this.Channel);
            if (this.MaxItems != 0)
                writer.writeInt32(5, this.MaxItems);
            if (this.WaitTimeout != 0)
                writer.writeInt32(6, this.WaitTimeout);
            if (this.AutoAck != false)
                writer.writeBool(7, this.AutoAck);
            if (this.ReQueueChannel.length)
                writer.writeString(8, this.ReQueueChannel);
            if (this.SequenceRange.length)
                writer.writePackedInt64(9, this.SequenceRange);
            if (this.RefTransactionId.length)
                writer.writeString(10, this.RefTransactionId);
            for (const [key, value] of this.Metadata) {
                writer.writeMessage(12, this.Metadata, () => {
                    writer.writeString(1, key);
                    writer.writeString(2, value);
                });
            }
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueuesDownstreamRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        message.ClientID = reader.readString();
                        break;
                    case 3:
                        message.RequestTypeData = reader.readEnum();
                        break;
                    case 4:
                        message.Channel = reader.readString();
                        break;
                    case 5:
                        message.MaxItems = reader.readInt32();
                        break;
                    case 6:
                        message.WaitTimeout = reader.readInt32();
                        break;
                    case 7:
                        message.AutoAck = reader.readBool();
                        break;
                    case 8:
                        message.ReQueueChannel = reader.readString();
                        break;
                    case 9:
                        message.SequenceRange = reader.readPackedInt64();
                        break;
                    case 10:
                        message.RefTransactionId = reader.readString();
                        break;
                    case 12:
                        reader.readMessage(message, () => pb_1.Map.deserializeBinary(message.Metadata, reader, reader.readString, reader.readString));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueuesDownstreamRequest.deserialize(bytes);
        }
    }
    _QueuesDownstreamRequest_one_of_decls = new WeakMap();
    kubemq.QueuesDownstreamRequest = QueuesDownstreamRequest;
    class QueuesDownstreamResponse extends pb_1.Message {
        constructor(data) {
            super();
            _QueuesDownstreamResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [4, 5], tslib_1.__classPrivateFieldGet(this, _QueuesDownstreamResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("TransactionId" in data && data.TransactionId != undefined) {
                    this.TransactionId = data.TransactionId;
                }
                if ("RefRequestId" in data && data.RefRequestId != undefined) {
                    this.RefRequestId = data.RefRequestId;
                }
                if ("RequestTypeData" in data && data.RequestTypeData != undefined) {
                    this.RequestTypeData = data.RequestTypeData;
                }
                if ("Messages" in data && data.Messages != undefined) {
                    this.Messages = data.Messages;
                }
                if ("ActiveOffsets" in data && data.ActiveOffsets != undefined) {
                    this.ActiveOffsets = data.ActiveOffsets;
                }
                if ("IsError" in data && data.IsError != undefined) {
                    this.IsError = data.IsError;
                }
                if ("Error" in data && data.Error != undefined) {
                    this.Error = data.Error;
                }
                if ("TransactionComplete" in data && data.TransactionComplete != undefined) {
                    this.TransactionComplete = data.TransactionComplete;
                }
                if ("Metadata" in data && data.Metadata != undefined) {
                    this.Metadata = data.Metadata;
                }
            }
            if (!this.Metadata)
                this.Metadata = new Map();
        }
        get TransactionId() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set TransactionId(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get RefRequestId() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set RefRequestId(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get RequestTypeData() {
            return pb_1.Message.getFieldWithDefault(this, 3, QueuesDownstreamRequestType.PollRequestTypeUnknown);
        }
        set RequestTypeData(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Messages() {
            return pb_1.Message.getRepeatedWrapperField(this, QueueMessage, 4);
        }
        set Messages(value) {
            pb_1.Message.setRepeatedWrapperField(this, 4, value);
        }
        get ActiveOffsets() {
            return pb_1.Message.getFieldWithDefault(this, 5, []);
        }
        set ActiveOffsets(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get IsError() {
            return pb_1.Message.getFieldWithDefault(this, 6, false);
        }
        set IsError(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get Error() {
            return pb_1.Message.getFieldWithDefault(this, 7, "");
        }
        set Error(value) {
            pb_1.Message.setField(this, 7, value);
        }
        get TransactionComplete() {
            return pb_1.Message.getFieldWithDefault(this, 8, false);
        }
        set TransactionComplete(value) {
            pb_1.Message.setField(this, 8, value);
        }
        get Metadata() {
            return pb_1.Message.getField(this, 9);
        }
        set Metadata(value) {
            pb_1.Message.setField(this, 9, value);
        }
        static fromObject(data) {
            const message = new QueuesDownstreamResponse({});
            if (data.TransactionId != null) {
                message.TransactionId = data.TransactionId;
            }
            if (data.RefRequestId != null) {
                message.RefRequestId = data.RefRequestId;
            }
            if (data.RequestTypeData != null) {
                message.RequestTypeData = data.RequestTypeData;
            }
            if (data.Messages != null) {
                message.Messages = data.Messages.map(item => QueueMessage.fromObject(item));
            }
            if (data.ActiveOffsets != null) {
                message.ActiveOffsets = data.ActiveOffsets;
            }
            if (data.IsError != null) {
                message.IsError = data.IsError;
            }
            if (data.Error != null) {
                message.Error = data.Error;
            }
            if (data.TransactionComplete != null) {
                message.TransactionComplete = data.TransactionComplete;
            }
            if (typeof data.Metadata == "object") {
                message.Metadata = new Map(Object.entries(data.Metadata));
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.TransactionId != null) {
                data.TransactionId = this.TransactionId;
            }
            if (this.RefRequestId != null) {
                data.RefRequestId = this.RefRequestId;
            }
            if (this.RequestTypeData != null) {
                data.RequestTypeData = this.RequestTypeData;
            }
            if (this.Messages != null) {
                data.Messages = this.Messages.map((item) => item.toObject());
            }
            if (this.ActiveOffsets != null) {
                data.ActiveOffsets = this.ActiveOffsets;
            }
            if (this.IsError != null) {
                data.IsError = this.IsError;
            }
            if (this.Error != null) {
                data.Error = this.Error;
            }
            if (this.TransactionComplete != null) {
                data.TransactionComplete = this.TransactionComplete;
            }
            if (this.Metadata != null) {
                data.Metadata = (Object.fromEntries)(this.Metadata);
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.TransactionId.length)
                writer.writeString(1, this.TransactionId);
            if (this.RefRequestId.length)
                writer.writeString(2, this.RefRequestId);
            if (this.RequestTypeData != QueuesDownstreamRequestType.PollRequestTypeUnknown)
                writer.writeEnum(3, this.RequestTypeData);
            if (this.Messages.length)
                writer.writeRepeatedMessage(4, this.Messages, (item) => item.serialize(writer));
            if (this.ActiveOffsets.length)
                writer.writePackedInt64(5, this.ActiveOffsets);
            if (this.IsError != false)
                writer.writeBool(6, this.IsError);
            if (this.Error.length)
                writer.writeString(7, this.Error);
            if (this.TransactionComplete != false)
                writer.writeBool(8, this.TransactionComplete);
            for (const [key, value] of this.Metadata) {
                writer.writeMessage(9, this.Metadata, () => {
                    writer.writeString(1, key);
                    writer.writeString(2, value);
                });
            }
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueuesDownstreamResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.TransactionId = reader.readString();
                        break;
                    case 2:
                        message.RefRequestId = reader.readString();
                        break;
                    case 3:
                        message.RequestTypeData = reader.readEnum();
                        break;
                    case 4:
                        reader.readMessage(message.Messages, () => pb_1.Message.addToRepeatedWrapperField(message, 4, QueueMessage.deserialize(reader), QueueMessage));
                        break;
                    case 5:
                        message.ActiveOffsets = reader.readPackedInt64();
                        break;
                    case 6:
                        message.IsError = reader.readBool();
                        break;
                    case 7:
                        message.Error = reader.readString();
                        break;
                    case 8:
                        message.TransactionComplete = reader.readBool();
                        break;
                    case 9:
                        reader.readMessage(message, () => pb_1.Map.deserializeBinary(message.Metadata, reader, reader.readString, reader.readString));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueuesDownstreamResponse.deserialize(bytes);
        }
    }
    _QueuesDownstreamResponse_one_of_decls = new WeakMap();
    kubemq.QueuesDownstreamResponse = QueuesDownstreamResponse;
    class QueueInfo extends pb_1.Message {
        constructor(data) {
            super();
            _QueueInfo_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _QueueInfo_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("Name" in data && data.Name != undefined) {
                    this.Name = data.Name;
                }
                if ("Messages" in data && data.Messages != undefined) {
                    this.Messages = data.Messages;
                }
                if ("Bytes" in data && data.Bytes != undefined) {
                    this.Bytes = data.Bytes;
                }
                if ("FirstSequence" in data && data.FirstSequence != undefined) {
                    this.FirstSequence = data.FirstSequence;
                }
                if ("LastSequence" in data && data.LastSequence != undefined) {
                    this.LastSequence = data.LastSequence;
                }
                if ("Sent" in data && data.Sent != undefined) {
                    this.Sent = data.Sent;
                }
                if ("Delivered" in data && data.Delivered != undefined) {
                    this.Delivered = data.Delivered;
                }
                if ("Waiting" in data && data.Waiting != undefined) {
                    this.Waiting = data.Waiting;
                }
                if ("Subscribers" in data && data.Subscribers != undefined) {
                    this.Subscribers = data.Subscribers;
                }
            }
        }
        get Name() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set Name(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Messages() {
            return pb_1.Message.getFieldWithDefault(this, 2, 0);
        }
        set Messages(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Bytes() {
            return pb_1.Message.getFieldWithDefault(this, 3, 0);
        }
        set Bytes(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get FirstSequence() {
            return pb_1.Message.getFieldWithDefault(this, 4, 0);
        }
        set FirstSequence(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get LastSequence() {
            return pb_1.Message.getFieldWithDefault(this, 5, 0);
        }
        set LastSequence(value) {
            pb_1.Message.setField(this, 5, value);
        }
        get Sent() {
            return pb_1.Message.getFieldWithDefault(this, 6, 0);
        }
        set Sent(value) {
            pb_1.Message.setField(this, 6, value);
        }
        get Delivered() {
            return pb_1.Message.getFieldWithDefault(this, 7, 0);
        }
        set Delivered(value) {
            pb_1.Message.setField(this, 7, value);
        }
        get Waiting() {
            return pb_1.Message.getFieldWithDefault(this, 8, 0);
        }
        set Waiting(value) {
            pb_1.Message.setField(this, 8, value);
        }
        get Subscribers() {
            return pb_1.Message.getFieldWithDefault(this, 9, 0);
        }
        set Subscribers(value) {
            pb_1.Message.setField(this, 9, value);
        }
        static fromObject(data) {
            const message = new QueueInfo({});
            if (data.Name != null) {
                message.Name = data.Name;
            }
            if (data.Messages != null) {
                message.Messages = data.Messages;
            }
            if (data.Bytes != null) {
                message.Bytes = data.Bytes;
            }
            if (data.FirstSequence != null) {
                message.FirstSequence = data.FirstSequence;
            }
            if (data.LastSequence != null) {
                message.LastSequence = data.LastSequence;
            }
            if (data.Sent != null) {
                message.Sent = data.Sent;
            }
            if (data.Delivered != null) {
                message.Delivered = data.Delivered;
            }
            if (data.Waiting != null) {
                message.Waiting = data.Waiting;
            }
            if (data.Subscribers != null) {
                message.Subscribers = data.Subscribers;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.Name != null) {
                data.Name = this.Name;
            }
            if (this.Messages != null) {
                data.Messages = this.Messages;
            }
            if (this.Bytes != null) {
                data.Bytes = this.Bytes;
            }
            if (this.FirstSequence != null) {
                data.FirstSequence = this.FirstSequence;
            }
            if (this.LastSequence != null) {
                data.LastSequence = this.LastSequence;
            }
            if (this.Sent != null) {
                data.Sent = this.Sent;
            }
            if (this.Delivered != null) {
                data.Delivered = this.Delivered;
            }
            if (this.Waiting != null) {
                data.Waiting = this.Waiting;
            }
            if (this.Subscribers != null) {
                data.Subscribers = this.Subscribers;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.Name.length)
                writer.writeString(1, this.Name);
            if (this.Messages != 0)
                writer.writeInt64(2, this.Messages);
            if (this.Bytes != 0)
                writer.writeInt64(3, this.Bytes);
            if (this.FirstSequence != 0)
                writer.writeInt64(4, this.FirstSequence);
            if (this.LastSequence != 0)
                writer.writeInt64(5, this.LastSequence);
            if (this.Sent != 0)
                writer.writeInt64(6, this.Sent);
            if (this.Delivered != 0)
                writer.writeInt64(7, this.Delivered);
            if (this.Waiting != 0)
                writer.writeInt64(8, this.Waiting);
            if (this.Subscribers != 0)
                writer.writeInt64(9, this.Subscribers);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueueInfo();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.Name = reader.readString();
                        break;
                    case 2:
                        message.Messages = reader.readInt64();
                        break;
                    case 3:
                        message.Bytes = reader.readInt64();
                        break;
                    case 4:
                        message.FirstSequence = reader.readInt64();
                        break;
                    case 5:
                        message.LastSequence = reader.readInt64();
                        break;
                    case 6:
                        message.Sent = reader.readInt64();
                        break;
                    case 7:
                        message.Delivered = reader.readInt64();
                        break;
                    case 8:
                        message.Waiting = reader.readInt64();
                        break;
                    case 9:
                        message.Subscribers = reader.readInt64();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueueInfo.deserialize(bytes);
        }
    }
    _QueueInfo_one_of_decls = new WeakMap();
    kubemq.QueueInfo = QueueInfo;
    class QueuesInfo extends pb_1.Message {
        constructor(data) {
            super();
            _QueuesInfo_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [5], tslib_1.__classPrivateFieldGet(this, _QueuesInfo_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("TotalQueue" in data && data.TotalQueue != undefined) {
                    this.TotalQueue = data.TotalQueue;
                }
                if ("Sent" in data && data.Sent != undefined) {
                    this.Sent = data.Sent;
                }
                if ("Delivered" in data && data.Delivered != undefined) {
                    this.Delivered = data.Delivered;
                }
                if ("Waiting" in data && data.Waiting != undefined) {
                    this.Waiting = data.Waiting;
                }
                if ("Queues" in data && data.Queues != undefined) {
                    this.Queues = data.Queues;
                }
            }
        }
        get TotalQueue() {
            return pb_1.Message.getFieldWithDefault(this, 1, 0);
        }
        set TotalQueue(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Sent() {
            return pb_1.Message.getFieldWithDefault(this, 2, 0);
        }
        set Sent(value) {
            pb_1.Message.setField(this, 2, value);
        }
        get Delivered() {
            return pb_1.Message.getFieldWithDefault(this, 3, 0);
        }
        set Delivered(value) {
            pb_1.Message.setField(this, 3, value);
        }
        get Waiting() {
            return pb_1.Message.getFieldWithDefault(this, 4, 0);
        }
        set Waiting(value) {
            pb_1.Message.setField(this, 4, value);
        }
        get Queues() {
            return pb_1.Message.getRepeatedWrapperField(this, QueueInfo, 5);
        }
        set Queues(value) {
            pb_1.Message.setRepeatedWrapperField(this, 5, value);
        }
        static fromObject(data) {
            const message = new QueuesInfo({});
            if (data.TotalQueue != null) {
                message.TotalQueue = data.TotalQueue;
            }
            if (data.Sent != null) {
                message.Sent = data.Sent;
            }
            if (data.Delivered != null) {
                message.Delivered = data.Delivered;
            }
            if (data.Waiting != null) {
                message.Waiting = data.Waiting;
            }
            if (data.Queues != null) {
                message.Queues = data.Queues.map(item => QueueInfo.fromObject(item));
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.TotalQueue != null) {
                data.TotalQueue = this.TotalQueue;
            }
            if (this.Sent != null) {
                data.Sent = this.Sent;
            }
            if (this.Delivered != null) {
                data.Delivered = this.Delivered;
            }
            if (this.Waiting != null) {
                data.Waiting = this.Waiting;
            }
            if (this.Queues != null) {
                data.Queues = this.Queues.map((item) => item.toObject());
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.TotalQueue != 0)
                writer.writeInt32(1, this.TotalQueue);
            if (this.Sent != 0)
                writer.writeInt64(2, this.Sent);
            if (this.Delivered != 0)
                writer.writeInt64(3, this.Delivered);
            if (this.Waiting != 0)
                writer.writeInt64(4, this.Waiting);
            if (this.Queues.length)
                writer.writeRepeatedMessage(5, this.Queues, (item) => item.serialize(writer));
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueuesInfo();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.TotalQueue = reader.readInt32();
                        break;
                    case 2:
                        message.Sent = reader.readInt64();
                        break;
                    case 3:
                        message.Delivered = reader.readInt64();
                        break;
                    case 4:
                        message.Waiting = reader.readInt64();
                        break;
                    case 5:
                        reader.readMessage(message.Queues, () => pb_1.Message.addToRepeatedWrapperField(message, 5, QueueInfo.deserialize(reader), QueueInfo));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueuesInfo.deserialize(bytes);
        }
    }
    _QueuesInfo_one_of_decls = new WeakMap();
    kubemq.QueuesInfo = QueuesInfo;
    class QueuesInfoRequest extends pb_1.Message {
        constructor(data) {
            super();
            _QueuesInfoRequest_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _QueuesInfoRequest_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RequestID" in data && data.RequestID != undefined) {
                    this.RequestID = data.RequestID;
                }
                if ("QueueName" in data && data.QueueName != undefined) {
                    this.QueueName = data.QueueName;
                }
            }
        }
        get RequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get QueueName() {
            return pb_1.Message.getFieldWithDefault(this, 2, "");
        }
        set QueueName(value) {
            pb_1.Message.setField(this, 2, value);
        }
        static fromObject(data) {
            const message = new QueuesInfoRequest({});
            if (data.RequestID != null) {
                message.RequestID = data.RequestID;
            }
            if (data.QueueName != null) {
                message.QueueName = data.QueueName;
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RequestID != null) {
                data.RequestID = this.RequestID;
            }
            if (this.QueueName != null) {
                data.QueueName = this.QueueName;
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RequestID.length)
                writer.writeString(1, this.RequestID);
            if (this.QueueName.length)
                writer.writeString(2, this.QueueName);
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueuesInfoRequest();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RequestID = reader.readString();
                        break;
                    case 2:
                        message.QueueName = reader.readString();
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueuesInfoRequest.deserialize(bytes);
        }
    }
    _QueuesInfoRequest_one_of_decls = new WeakMap();
    kubemq.QueuesInfoRequest = QueuesInfoRequest;
    class QueuesInfoResponse extends pb_1.Message {
        constructor(data) {
            super();
            _QueuesInfoResponse_one_of_decls.set(this, []);
            pb_1.Message.initialize(this, Array.isArray(data) ? data : [], 0, -1, [], tslib_1.__classPrivateFieldGet(this, _QueuesInfoResponse_one_of_decls, "f"));
            if (!Array.isArray(data) && typeof data == "object") {
                if ("RefRequestID" in data && data.RefRequestID != undefined) {
                    this.RefRequestID = data.RefRequestID;
                }
                if ("Info" in data && data.Info != undefined) {
                    this.Info = data.Info;
                }
            }
        }
        get RefRequestID() {
            return pb_1.Message.getFieldWithDefault(this, 1, "");
        }
        set RefRequestID(value) {
            pb_1.Message.setField(this, 1, value);
        }
        get Info() {
            return pb_1.Message.getWrapperField(this, QueuesInfo, 2);
        }
        set Info(value) {
            pb_1.Message.setWrapperField(this, 2, value);
        }
        get has_Info() {
            return pb_1.Message.getField(this, 2) != null;
        }
        static fromObject(data) {
            const message = new QueuesInfoResponse({});
            if (data.RefRequestID != null) {
                message.RefRequestID = data.RefRequestID;
            }
            if (data.Info != null) {
                message.Info = QueuesInfo.fromObject(data.Info);
            }
            return message;
        }
        toObject() {
            const data = {};
            if (this.RefRequestID != null) {
                data.RefRequestID = this.RefRequestID;
            }
            if (this.Info != null) {
                data.Info = this.Info.toObject();
            }
            return data;
        }
        serialize(w) {
            const writer = w || new pb_1.BinaryWriter();
            if (this.RefRequestID.length)
                writer.writeString(1, this.RefRequestID);
            if (this.has_Info)
                writer.writeMessage(2, this.Info, () => this.Info.serialize(writer));
            if (!w)
                return writer.getResultBuffer();
        }
        static deserialize(bytes) {
            const reader = bytes instanceof pb_1.BinaryReader ? bytes : new pb_1.BinaryReader(bytes), message = new QueuesInfoResponse();
            while (reader.nextField()) {
                if (reader.isEndGroup())
                    break;
                switch (reader.getFieldNumber()) {
                    case 1:
                        message.RefRequestID = reader.readString();
                        break;
                    case 2:
                        reader.readMessage(message.Info, () => message.Info = QueuesInfo.deserialize(reader));
                        break;
                    default: reader.skipField();
                }
            }
            return message;
        }
        serializeBinary() {
            return this.serialize();
        }
        static deserializeBinary(bytes) {
            return QueuesInfoResponse.deserialize(bytes);
        }
    }
    _QueuesInfoResponse_one_of_decls = new WeakMap();
    kubemq.QueuesInfoResponse = QueuesInfoResponse;
    class UnimplementedkubemqService {
    }
    UnimplementedkubemqService.definition = {
        SendEvent: {
            path: "/kubemq.kubemq/SendEvent",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => Event.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => Result.deserialize(new Uint8Array(bytes))
        },
        SendEventsStream: {
            path: "/kubemq.kubemq/SendEventsStream",
            requestStream: true,
            responseStream: true,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => Event.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => Result.deserialize(new Uint8Array(bytes))
        },
        SubscribeToEvents: {
            path: "/kubemq.kubemq/SubscribeToEvents",
            requestStream: false,
            responseStream: true,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => Subscribe.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => EventReceive.deserialize(new Uint8Array(bytes))
        },
        SubscribeToRequests: {
            path: "/kubemq.kubemq/SubscribeToRequests",
            requestStream: false,
            responseStream: true,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => Subscribe.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => Request.deserialize(new Uint8Array(bytes))
        },
        SendRequest: {
            path: "/kubemq.kubemq/SendRequest",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => Request.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => Response.deserialize(new Uint8Array(bytes))
        },
        SendResponse: {
            path: "/kubemq.kubemq/SendResponse",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => Response.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => Empty.deserialize(new Uint8Array(bytes))
        },
        SendQueueMessage: {
            path: "/kubemq.kubemq/SendQueueMessage",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => QueueMessage.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => SendQueueMessageResult.deserialize(new Uint8Array(bytes))
        },
        SendQueueMessagesBatch: {
            path: "/kubemq.kubemq/SendQueueMessagesBatch",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => QueueMessagesBatchRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => QueueMessagesBatchResponse.deserialize(new Uint8Array(bytes))
        },
        ReceiveQueueMessages: {
            path: "/kubemq.kubemq/ReceiveQueueMessages",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => ReceiveQueueMessagesRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => ReceiveQueueMessagesResponse.deserialize(new Uint8Array(bytes))
        },
        StreamQueueMessage: {
            path: "/kubemq.kubemq/StreamQueueMessage",
            requestStream: true,
            responseStream: true,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => StreamQueueMessagesRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => StreamQueueMessagesResponse.deserialize(new Uint8Array(bytes))
        },
        AckAllQueueMessages: {
            path: "/kubemq.kubemq/AckAllQueueMessages",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => AckAllQueueMessagesRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => AckAllQueueMessagesResponse.deserialize(new Uint8Array(bytes))
        },
        Ping: {
            path: "/kubemq.kubemq/Ping",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => Empty.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => PingResult.deserialize(new Uint8Array(bytes))
        },
        QueuesDownstream: {
            path: "/kubemq.kubemq/QueuesDownstream",
            requestStream: true,
            responseStream: true,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => QueuesDownstreamRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => QueuesDownstreamResponse.deserialize(new Uint8Array(bytes))
        },
        QueuesUpstream: {
            path: "/kubemq.kubemq/QueuesUpstream",
            requestStream: true,
            responseStream: true,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => QueuesUpstreamRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => QueuesUpstreamResponse.deserialize(new Uint8Array(bytes))
        },
        QueuesInfo: {
            path: "/kubemq.kubemq/QueuesInfo",
            requestStream: false,
            responseStream: false,
            requestSerialize: (message) => Buffer.from(message.serialize()),
            requestDeserialize: (bytes) => QueuesInfoRequest.deserialize(new Uint8Array(bytes)),
            responseSerialize: (message) => Buffer.from(message.serialize()),
            responseDeserialize: (bytes) => QueuesInfoResponse.deserialize(new Uint8Array(bytes))
        }
    };
    kubemq.UnimplementedkubemqService = UnimplementedkubemqService;
    class kubemqClient extends grpc_1.makeGenericClientConstructor(UnimplementedkubemqService.definition, "kubemq", {}) {
        constructor(address, credentials, options) {
            super(address, credentials, options);
            this.SendEvent = (message, metadata, options, callback) => {
                return super.SendEvent(message, metadata, options, callback);
            };
            this.SendEventsStream = (metadata, options) => {
                return super.SendEventsStream(metadata, options);
            };
            this.SubscribeToEvents = (message, metadata, options) => {
                return super.SubscribeToEvents(message, metadata, options);
            };
            this.SubscribeToRequests = (message, metadata, options) => {
                return super.SubscribeToRequests(message, metadata, options);
            };
            this.SendRequest = (message, metadata, options, callback) => {
                return super.SendRequest(message, metadata, options, callback);
            };
            this.SendResponse = (message, metadata, options, callback) => {
                return super.SendResponse(message, metadata, options, callback);
            };
            this.SendQueueMessage = (message, metadata, options, callback) => {
                return super.SendQueueMessage(message, metadata, options, callback);
            };
            this.SendQueueMessagesBatch = (message, metadata, options, callback) => {
                return super.SendQueueMessagesBatch(message, metadata, options, callback);
            };
            this.ReceiveQueueMessages = (message, metadata, options, callback) => {
                return super.ReceiveQueueMessages(message, metadata, options, callback);
            };
            this.StreamQueueMessage = (metadata, options) => {
                return super.StreamQueueMessage(metadata, options);
            };
            this.AckAllQueueMessages = (message, metadata, options, callback) => {
                return super.AckAllQueueMessages(message, metadata, options, callback);
            };
            this.Ping = (message, metadata, options, callback) => {
                return super.Ping(message, metadata, options, callback);
            };
            this.QueuesDownstream = (metadata, options) => {
                return super.QueuesDownstream(metadata, options);
            };
            this.QueuesUpstream = (metadata, options) => {
                return super.QueuesUpstream(metadata, options);
            };
            this.QueuesInfo = (message, metadata, options, callback) => {
                return super.QueuesInfo(message, metadata, options, callback);
            };
        }
    }
    kubemq.kubemqClient = kubemqClient;
})(kubemq || (exports.kubemq = kubemq = {}));
